<!DOCTYPE html>
<html>
	<head>
		<meta charset="UTF-8">
<meta http-equiv="x-ua-compatible" content="ie=edge">
<meta name="viewport" content="width=device-width, minimum-scale=1.0">
<title>Intro R</title>

	<meta name="description" content="">


<link rel="icon" href="/favicon.ico" type="image/x-icon">
<link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Montserrat:400,400i,500,500i,700,700i|Noto+Sans:400,400i,700,700i|Source+Code+Pro&amp;subset=latin-ext">
<link rel="stylesheet" href="/assets/css/custom_styles.css">
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0/css/bootstrap.min.css" integrity="sha384-Gn5384xqQ1aoWXA+058RXPxPg6fy4IWvTNh0E263XmFcJlSAwiGgFAW/dAiS6JXm" crossorigin="anonymous">
<link rel="stylesheet" href="/doks-theme/assets/css/style.css">

	</head>
	<body class="green" data-spy="scroll" data-target=".js-scrollspy">
		
	<div class="offcanvas visible-xs">
		<ul class="offcanvas__nav">
			
				<li><a href="/">Home</a></li>
			
				<li><a href="/schedule">Schedule</a></li>
			
		</ul><!-- /.offcanvas__nav -->
	</div><!-- /.offcanvas -->



	<header class="site-header">
      <!--  <a href="https://github.com/qosf/qosf.org"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://s3.amazonaws.com/github/ribbons/forkme_right_white_ffffff.png" alt="Fork me on GitHub"></a>
 -->
		<div class="container">
			<div class="row">
				<div class="col-xs-12">
					
						<a href="/" class="site-header__logo">
                            <img src="/assets/img/logos/qosf_colour_logo.svg">
						</a><!-- /.site-header__logo -->
					
					
						<ul class="site-header__nav hidden-xs">
							
								<li><a href="/">Home</a></li>
							
								<li><a href="/schedule">Schedule</a></li>
							
						</ul><!-- /.site-header__nav -->
						<button class="offcanvas-toggle visible-xs">
							<span></span>
							<span></span>
							<span></span>
						</button><!-- /.offcanvas-toggle -->
					
				</div><!-- /.col -->
			</div><!-- /.row -->
		</div><!-- /.container -->
	</header><!-- /.site-header -->


		<div class="hero-subheader">
            
            
            
			<div class="container">
				<div class="row">
					<div class="col-md-7">
						<div class="align-container" data-mh>
							<div class="align-inner">
								
									<h1 class="hero-subheader__title">Intro R</h1>
								
								
                                <p class="hero-subheader__desc">
</p>
								
                                
							</div><!-- /.align-inner -->
						</div><!-- /.align-container -->
					</div><!-- /.col -->
					
				</div><!-- /.row -->
			</div><!-- /.container -->
		</div><!-- /.hero-subheader -->
		<div class="section">
			<div class="container">
				<div class="row">
					<div class="col-md-7">
						<div class="content">
							<h1 id="introduction-to-r">Introduction to R</h1>

<h2 id="rstudio">RStudio</h2>

<p>Some userful shortcuts: <code class="language-plaintext highlighter-rouge">Ctrl+Enter</code> (send code), <code class="language-plaintext highlighter-rouge">Ctrl+Shift+C</code>
(comment/uncomment), <code class="language-plaintext highlighter-rouge">Ctrl+1/2</code> (switch window focus)</p>

<p>Explore <a href="https://rstudio.com/wp-content/uploads/2016/01/rstudio-IDE-cheatsheet.pdf">here</a> all the RStudio shortcuts</p>

<h2 id="r-package-repositories">R Package Repositories</h2>

<ul>
  <li>CRAN (&gt;11,000 packages) general data analysis -
  <a href="http://cran.at.r-project.org/">URL</a></li>
  <li>Bioconductor (&gt;1,800 packages) bioscience data analysis -
  <a href="http://www.bioconductor.org/">URL</a></li>
  <li>Omegahat (&gt;90 packages) programming interfaces -
  <a href="https://github.com/omegahat?tab=repositories">URL</a></li>
</ul>

<h2 id="installation-of-r-packages">Installation of R Packages</h2>

<ol>
  <li>
    <p>Install R for your operating system from
<a href="http://cran.at.r-project.org/">CRAN</a>.</p>
  </li>
  <li>
    <p>Install RStudio from <a href="http://www.rstudio.com/ide/download">RStudio</a>.</p>
  </li>
  <li>
    <p>Install CRAN Packages from R console like this:</p>
  </li>
</ol>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>install.packages(c("pkg1", "pkg2")) 
install.packages("pkg.zip", repos=NULL)
</code></pre></div></div>

<ol>
  <li>Install Bioconductor packages as follows:</li>
</ol>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>install.packages("BiocManager") 
BiocManager::install("systemPipeR")
</code></pre></div></div>

<h2 id="getting-around">Getting Around</h2>

<h3 id="startup-and-closing-behavior">Startup and Closing Behavior</h3>

<ul>
  <li>
    <p><strong>Starting R</strong>: The R GUI versions, including RStudio, under Windows
  and Mac OS X can be opened by double-clicking their icons.
  Alternatively, one can start it by typing <code class="language-plaintext highlighter-rouge">R</code> in a terminal (default
  under Linux).</p>
  </li>
  <li>
    <p><strong>Startup/Closing Behavior</strong>: The R environment is controlled by
  hidden files in the startup directory: <code class="language-plaintext highlighter-rouge">.RData</code>, <code class="language-plaintext highlighter-rouge">.Rhistory</code> and
  <code class="language-plaintext highlighter-rouge">.Rprofile</code> (optional).</p>
  </li>
  <li>
    <p><strong>Closing R</strong>:</p>
  </li>
</ul>

<p>```{r closing_r, eval=FALSE}
q()</p>
<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code></code></pre></div></div>
<p>Save workspace image? [y/n/c]:
```
<!-- --></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>q()  

Save workspace image? [y/n/c]:
</code></pre></div></div>

<ul>
  <li><strong>Note</strong>: When responding with <code class="language-plaintext highlighter-rouge">y</code>, then the entire R workspace will
  be written to the <code class="language-plaintext highlighter-rouge">.RData</code> file which can become very large. Often
  it is sufficient to just save an analysis protocol in an R source
  file. This way one can quickly regenerate all data sets and objects.</li>
</ul>

<h3 id="navigating-directories">Navigating directories</h3>

<p>Create an object with the assignment operator <code class="language-plaintext highlighter-rouge">&lt;-</code> or <code class="language-plaintext highlighter-rouge">=</code></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>object &lt;- ...
</code></pre></div></div>

<p>List objects in current R session</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ls()
</code></pre></div></div>

<p>Return content of current working directory</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>dir()
</code></pre></div></div>

<p>Return path of current working directory</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>getwd()
</code></pre></div></div>

<p>Change current working directory</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>setwd("/home/user")
</code></pre></div></div>

<h2 id="basic-syntax">Basic Syntax</h2>

<p>General R command syntax</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>object &lt;- function_name(arguments) 
object &lt;- object[arguments] 
</code></pre></div></div>

<p>Finding help</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>?function_name
</code></pre></div></div>

<p>Load a library/package</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library("my_library") 
</code></pre></div></div>

<p>List functions defined by a library</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library(help="my_library")
</code></pre></div></div>

<p>Load library manual (PDF or HTML file)</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>vignette("my_library") 
</code></pre></div></div>

<p>Execute an R script from within R</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>source("my_script.R")
</code></pre></div></div>

<p>Execute an R script from command-line (the first of the three options is
preferred)</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ Rscript my_script.R
$ R CMD BATCH my_script.R 
$ R --slave &lt; my_script.R 
</code></pre></div></div>

<h2 id="data-types">Data Types</h2>

<h3 id="numeric-data">Numeric data</h3>

<p>Example: <code class="language-plaintext highlighter-rouge">1, 2, 3, ...</code></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>x &lt;- c(1, 2, 3)
x

## [1] 1 2 3

is.numeric(x)

## [1] TRUE

as.character(x)

## [1] "1" "2" "3"
</code></pre></div></div>

<h3 id="character-data">Character data</h3>

<p>Example: <code class="language-plaintext highlighter-rouge">"a", "b", "c", ...</code></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>x &lt;- c("1", "2", "3")
x

## [1] "1" "2" "3"

is.character(x)

## [1] TRUE

as.numeric(x)

## [1] 1 2 3
</code></pre></div></div>

<h3 id="complex-data">Complex data</h3>

<p>Example: mix of both</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>c(1, "b", 3)

## [1] "1" "b" "3"
</code></pre></div></div>

<h3 id="logical-data">Logical data</h3>

<p>Example: <code class="language-plaintext highlighter-rouge">TRUE</code> of <code class="language-plaintext highlighter-rouge">FALSE</code></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>x &lt;- 1:10 &lt; 5
x  

##  [1]  TRUE  TRUE  TRUE  TRUE FALSE FALSE FALSE FALSE FALSE FALSE

!x

##  [1] FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE  TRUE

which(x) # Returns index for the 'TRUE' values in logical vector

## [1] 1 2 3 4
</code></pre></div></div>

<h2 id="data-objects">Data Objects</h2>

<h3 id="object-types">Object types</h3>

<h3 id="vectors-1d">Vectors (1D)</h3>

<p>Definition: <code class="language-plaintext highlighter-rouge">numeric</code> or <code class="language-plaintext highlighter-rouge">character</code></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>myVec &lt;- 1:10; names(myVec) &lt;- letters[1:10]  
myVec[1:5]

## a b c d e 
## 1 2 3 4 5

myVec[c(2,4,6,8)]

## b d f h 
## 2 4 6 8

myVec[c("b", "d", "f")]

## b d f 
## 2 4 6
</code></pre></div></div>

<h3 id="factors-1d">Factors (1D)</h3>

<p>Definition: vectors with grouping information</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>factor(c("dog", "cat", "mouse", "dog", "dog", "cat"))

## [1] dog   cat   mouse dog   dog   cat  
## Levels: cat dog mouse
</code></pre></div></div>

<h3 id="matrices-2d">Matrices (2D)</h3>

<p>Definition: two dimensional structures with data of same type</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>myMA &lt;- matrix(1:30, 3, 10, byrow = TRUE) 
class(myMA)

## [1] "matrix" "array"

myMA[1:2,]

##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## [1,]    1    2    3    4    5    6    7    8    9    10
## [2,]   11   12   13   14   15   16   17   18   19    20

myMA[1, , drop=FALSE]

##      [,1] [,2] [,3] [,4] [,5] [,6] [,7] [,8] [,9] [,10]
## [1,]    1    2    3    4    5    6    7    8    9    10
</code></pre></div></div>

<h3 id="data-frames-2d">Data Frames (2D)</h3>

<p>Definition: two dimensional objects with data of variable types</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>myDF &lt;- data.frame(Col1=1:10, Col2=10:1) 
myDF[1:2, ]

##   Col1 Col2
## 1    1   10
## 2    2    9
</code></pre></div></div>

<h3 id="arrays">Arrays</h3>

<p>Definition: data structure with one, two or more dimensions</p>

<h3 id="lists">Lists</h3>

<p>Definition: containers for any object type</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>myL &lt;- list(name="Fred", wife="Mary", no.children=3, child.ages=c(4,7,9)) 
myL

## $name
## [1] "Fred"
## 
## $wife
## [1] "Mary"
## 
## $no.children
## [1] 3
## 
## $child.ages
## [1] 4 7 9

myL[[4]][1:2] 

## [1] 4 7
</code></pre></div></div>

<h3 id="functions">Functions</h3>

<p>Definition: piece of code</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>myfct &lt;- function(arg1, arg2, ...) { 
    function_body 
}
</code></pre></div></div>

<h3 id="subsetting-of-data-objects">Subsetting of data objects</h3>

<p><strong>(1.) Subsetting by positive or negative index/position numbers</strong></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>myVec &lt;- 1:26; names(myVec) &lt;- LETTERS 
myVec[1:4]

## A B C D 
## 1 2 3 4
</code></pre></div></div>

<p><strong>(2.) Subsetting by same length logical vectors</strong></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>myLog &lt;- myVec &gt; 10
myVec[myLog] 

##  K  L  M  N  O  P  Q  R  S  T  U  V  W  X  Y  Z 
## 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26
</code></pre></div></div>

<p><strong>(3.) Subsetting by field names</strong></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>myVec[c("B", "K", "M")]

##  B  K  M 
##  2 11 13
</code></pre></div></div>

<p><strong>(4.) Subset with <code class="language-plaintext highlighter-rouge">$</code> sign</strong>: references a single column or list
component by its name</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>iris$Species[1:8]

## [1] setosa setosa setosa setosa setosa setosa setosa setosa
## Levels: setosa versicolor virginica
</code></pre></div></div>

<h2 id="important-utilities">Important Utilities</h2>

<h3 id="combining-objects">Combining Objects</h3>

<p>The <code class="language-plaintext highlighter-rouge">c</code> function combines vectors and lists</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>c(1, 2, 3)

## [1] 1 2 3

x &lt;- 1:3; y &lt;- 101:103
c(x, y)

## [1]   1   2   3 101 102 103

iris$Species[1:8]

## [1] setosa setosa setosa setosa setosa setosa setosa setosa
## Levels: setosa versicolor virginica
</code></pre></div></div>

<p>The <code class="language-plaintext highlighter-rouge">cbind</code> and <code class="language-plaintext highlighter-rouge">rbind</code> functions can be used to append columns and
rows, respecively.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ma &lt;- cbind(x, y)
ma

##      x   y
## [1,] 1 101
## [2,] 2 102
## [3,] 3 103

rbind(ma, ma)

##      x   y
## [1,] 1 101
## [2,] 2 102
## [3,] 3 103
## [4,] 1 101
## [5,] 2 102
## [6,] 3 103
</code></pre></div></div>

<h3 id="accessing-dimensions-of-objects">Accessing Dimensions of Objects</h3>

<p>Length and dimension information of objects</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>length(iris$Species)

## [1] 150

dim(iris)

## [1] 150   5
</code></pre></div></div>

<h3 id="accessing-name-slots-of-objects">Accessing Name Slots of Objects</h3>

<p>Accessing row and column names of 2D objects</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>rownames(iris)[1:8]

## [1] "1" "2" "3" "4" "5" "6" "7" "8"

colnames(iris)

## [1] "Sepal.Length" "Sepal.Width"  "Petal.Length" "Petal.Width"  "Species"
</code></pre></div></div>

<p>Return name field of vectors and lists</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>names(myVec)

##  [1] "A" "B" "C" "D" "E" "F" "G" "H" "I" "J" "K" "L" "M" "N" "O" "P" "Q" "R" "S" "T" "U" "V" "W" "X"
## [25] "Y" "Z"

names(myL)

## [1] "name"        "wife"        "no.children" "child.ages"
</code></pre></div></div>

<h3 id="sorting-objects">Sorting Objects</h3>

<p>The function <code class="language-plaintext highlighter-rouge">sort</code> returns a vector in ascending or descending order</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sort(10:1)

##  [1]  1  2  3  4  5  6  7  8  9 10
</code></pre></div></div>

<p>The function <code class="language-plaintext highlighter-rouge">order</code> returns a sorting index for sorting an object</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sortindex &lt;- order(iris[,1], decreasing = FALSE)
sortindex[1:12]

##  [1] 14  9 39 43 42  4  7 23 48  3 30 12

iris[sortindex,][1:2,]

##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 14          4.3         3.0          1.1         0.1  setosa
## 9           4.4         2.9          1.4         0.2  setosa

sortindex &lt;- order(-iris[,1]) # Same as decreasing=TRUE
</code></pre></div></div>

<p>Sorting multiple columns</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>iris[order(iris$Sepal.Length, iris$Sepal.Width),][1:2,]

##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 14          4.3         3.0          1.1         0.1  setosa
## 9           4.4         2.9          1.4         0.2  setosa
</code></pre></div></div>

<h2 id="operators-and-calculations">Operators and Calculations</h2>

<h3 id="comparison-operators">Comparison Operators</h3>

<p>Comparison operators: <code class="language-plaintext highlighter-rouge">==</code>, <code class="language-plaintext highlighter-rouge">!=</code>, <code class="language-plaintext highlighter-rouge">&lt;</code>, <code class="language-plaintext highlighter-rouge">&gt;</code>, <code class="language-plaintext highlighter-rouge">&lt;=</code>, <code class="language-plaintext highlighter-rouge">&gt;=</code></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>1==1

## [1] TRUE
</code></pre></div></div>

<p>Logical operators: AND: <code class="language-plaintext highlighter-rouge">&amp;</code>, OR: <code class="language-plaintext highlighter-rouge">|</code>, NOT: <code class="language-plaintext highlighter-rouge">!</code></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>x &lt;- 1:10; y &lt;- 10:1
x &gt; y &amp; x &gt; 5

##  [1] FALSE FALSE FALSE FALSE FALSE  TRUE  TRUE  TRUE  TRUE  TRUE
</code></pre></div></div>

<h3 id="basic-calculations">Basic Calculations</h3>

<p>To look up math functions, see Function Index
<a href="http://cran.at.r-project.org/doc/manuals/R-intro.html#Function-and-variable-index">here</a></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>x + y

##  [1] 11 11 11 11 11 11 11 11 11 11

sum(x)

## [1] 55

mean(x)

## [1] 5.5

apply(iris[1:6,1:3], 1, mean) 

##        1        2        3        4        5        6 
## 3.333333 3.100000 3.066667 3.066667 3.333333 3.666667
</code></pre></div></div>

<h2 id="reading-and-writing-external-data">Reading and Writing External Data</h2>

<h3 id="import-of-tabular-data">Import of tabular data</h3>

<p>Import of a tab-delimited tabular file</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>myDF &lt;- read.delim("myData.xls", sep="\t")
</code></pre></div></div>

<p>Import of Excel file. Note: working with tab- or comma-delimited files
is more flexible and preferred.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library(gdata)
myDF &lt;- read.xls"myData.xls")
</code></pre></div></div>

<p>Import of Google Sheets. The following example imports a sample Google
Sheet from
<a href="https://docs.google.com/spreadsheets/d/1U-32UcwZP1k3saKeaH1mbvEAOfZRdNHNkWK2GI1rpPM/edit#gid=472150521">here</a>.
Detailed instructions for interacting from R with Google Sheets with the
required <code class="language-plaintext highlighter-rouge">googlesheets</code> package are
<a href="https://github.com/jennybc/googlesheets">here</a>.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library("googlesheets"); library("dplyr"); library(knitr)
gs_auth() # Creates authorizaton token (.httr-oauth) in current directory if not present
sheetid &lt;-"1U-32UcwZP1k3saKeaH1mbvEAOfZRdNHNkWK2GI1rpPM"
gap &lt;- gs_key(sheetid)
mysheet &lt;- gs_read(gap, skip=4)
myDF &lt;- as.data.frame(mysheet)
myDF
</code></pre></div></div>

<h3 id="export-of-tabular-data">Export of tabular data</h3>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>write.table(myDF, file="myfile.xls", sep="\t", quote=FALSE, col.names=NA)
</code></pre></div></div>

<h3 id="line-wise-import">Line-wise import</h3>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>myDF &lt;- readLines("myData.txt")
</code></pre></div></div>

<h3 id="line-wise-export">Line-wise export</h3>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>writeLines(month.name, "myData.txt")
</code></pre></div></div>

<h3 id="export-r-object">Export R object</h3>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mylist &lt;- list(C1=iris[,1], C2=iris[,2]) # Example to export
saveRDS(mylist, "mylist.rds")
</code></pre></div></div>

<h3 id="import-r-object">Import R object</h3>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mylist &lt;- readRDS("mylist.rds")
</code></pre></div></div>

<h3 id="copy-and-paste-into-r">Copy and paste into R</h3>

<p>On Windows/Linux systems</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>read.delim("clipboard") 
</code></pre></div></div>

<p>On Mac OS X systems</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>read.delim(pipe("pbpaste")) 
</code></pre></div></div>

<h3 id="copy-and-paste-from-r">Copy and paste from R</h3>

<p>On Windows/Linux systems</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>write.table(iris, "clipboard", sep="\t", col.names=NA, quote=F) 
</code></pre></div></div>

<p>On Mac OS X systems</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>zz &lt;- pipe('pbcopy', 'w')
write.table(iris, zz, sep="\t", col.names=NA, quote=F)
close(zz) 
</code></pre></div></div>

<h3 id="homework-3a">Homework 3A</h3>

<p>Homework 3A: <a href="http://girke.bioinformatics.ucr.edu/GEN242/mydoc_homework_03.html">Object Subsetting Routines and
Import/Export</a></p>

<h2 id="useful-r-functions">Useful R Functions</h2>

<h3 id="unique-entries">Unique entries</h3>

<p>Make vector entries unique with <code class="language-plaintext highlighter-rouge">unique</code></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>length(iris$Sepal.Length)

## [1] 150

length(unique(iris$Sepal.Length))

## [1] 35
</code></pre></div></div>

<h3 id="count-occurrences">Count occurrences</h3>

<p>Count occurrences of entries with <code class="language-plaintext highlighter-rouge">table</code></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>table(iris$Species)

## 
##     setosa versicolor  virginica 
##         50         50         50
</code></pre></div></div>

<h3 id="aggregate-data">Aggregate data</h3>

<p>Compute aggregate statistics with <code class="language-plaintext highlighter-rouge">aggregate</code></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>aggregate(iris[,1:4], by=list(iris$Species), FUN=mean, na.rm=TRUE)

##      Group.1 Sepal.Length Sepal.Width Petal.Length Petal.Width
## 1     setosa        5.006       3.428        1.462       0.246
## 2 versicolor        5.936       2.770        4.260       1.326
## 3  virginica        6.588       2.974        5.552       2.026
</code></pre></div></div>

<h3 id="intersect-data">Intersect data</h3>

<p>Compute intersect between two vectors with <code class="language-plaintext highlighter-rouge">%in%</code></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>month.name %in% c("May", "July")

##  [1] FALSE FALSE FALSE FALSE  TRUE FALSE  TRUE FALSE FALSE FALSE FALSE FALSE
</code></pre></div></div>

<h3 id="merge-data-frames">Merge data frames</h3>

<p>Join two data frames by common field entries with <code class="language-plaintext highlighter-rouge">merge</code> (here row
names <code class="language-plaintext highlighter-rouge">by.x=0</code>). To obtain only the common rows, change <code class="language-plaintext highlighter-rouge">all=TRUE</code> to
<code class="language-plaintext highlighter-rouge">all=FALSE</code>. To merge on specific columns, refer to them by their
position numbers or their column names.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>frame1 &lt;- iris[sample(1:length(iris[,1]), 30), ]
frame1[1:2,]

##     Sepal.Length Sepal.Width Petal.Length Petal.Width    Species
## 127          6.2         2.8          4.8         1.8  virginica
## 52           6.4         3.2          4.5         1.5 versicolor

dim(frame1)

## [1] 30  5

my_result &lt;- merge(frame1, iris, by.x = 0, by.y = 0, all = TRUE)
dim(my_result)

## [1] 150  11
</code></pre></div></div>

<h2 id="dplyr-environment">dplyr Environment</h2>

<p>Modern object classes and methods for handling <code class="language-plaintext highlighter-rouge">data.frame</code> like
structures are provided by the <code class="language-plaintext highlighter-rouge">dplyr</code> and <code class="language-plaintext highlighter-rouge">data.table</code> packages. The
following gives a short introduction to the usage and functionalities of
the <code class="language-plaintext highlighter-rouge">dplyr</code> package. More detailed tutorials on this topic can be found
here:</p>

<ul>
  <li><a href="https://rdrr.io/cran/dplyr/">dplyr: A Grammar of Data Manipulation</a></li>
  <li><a href="https://cran.r-project.org/web/packages/dplyr/vignettes/dplyr.html">Introduction to
  <code class="language-plaintext highlighter-rouge">dplyr</code></a></li>
  <li><a href="http://genomicsclass.github.io/book/pages/dplyr_tutorial.html">Tutorial on
  <code class="language-plaintext highlighter-rouge">dplyr</code></a></li>
  <li><a href="http://stat545.com/bit001_dplyr-cheatsheet.html">Cheatsheet for Joins from Jenny
  Bryan</a></li>
  <li><a href="https://cran.r-project.org/web/packages/tibble/vignettes/tibble.html">Tibbles</a></li>
  <li><a href="https://www.r-bloggers.com/intro-to-the-data-table-package/">Intro to <code class="language-plaintext highlighter-rouge">data.table</code>
  package</a></li>
  <li><a href="https://www.r-bloggers.com/working-with-large-datasets-with-dplyr-and-data-table/">Big data with <code class="language-plaintext highlighter-rouge">dplyr</code> and
  <code class="language-plaintext highlighter-rouge">data.table</code></a></li>
  <li><a href="https://www.r-bloggers.com/fast-data-lookups-in-r-dplyr-vs-data-table/">Fast lookups with <code class="language-plaintext highlighter-rouge">dplyr</code> and
  <code class="language-plaintext highlighter-rouge">data.table</code></a></li>
</ul>

<h3 id="installation">Installation</h3>

<p>The <code class="language-plaintext highlighter-rouge">dplyr</code> environment has evolved into an ecosystem of packages. To
simplify package management, one can install and load the entire
collection via the <code class="language-plaintext highlighter-rouge">tidyverse</code> package. For more details on <code class="language-plaintext highlighter-rouge">tidyverse</code>
see <a href="http://tidyverse.org/">here</a>.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>install.packages("tidyverse")
</code></pre></div></div>

<h3 id="construct-a-tibble-tibble">Construct a <code class="language-plaintext highlighter-rouge">tibble</code> (<code class="language-plaintext highlighter-rouge">tibble</code>)</h3>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library(tidyverse)
as_tibble(iris) # coerce data.frame to tibble tbl

## # A tibble: 150 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;fct&gt;  
##  1          5.1         3.5          1.4         0.2 setosa 
##  2          4.9         3            1.4         0.2 setosa 
##  3          4.7         3.2          1.3         0.2 setosa 
##  4          4.6         3.1          1.5         0.2 setosa 
##  5          5           3.6          1.4         0.2 setosa 
##  6          5.4         3.9          1.7         0.4 setosa 
##  7          4.6         3.4          1.4         0.3 setosa 
##  8          5           3.4          1.5         0.2 setosa 
##  9          4.4         2.9          1.4         0.2 setosa 
## 10          4.9         3.1          1.5         0.1 setosa 
## # … with 140 more rows
</code></pre></div></div>

<p>Alternative functions producing the same result include <code class="language-plaintext highlighter-rouge">as_data_frame</code>
and <code class="language-plaintext highlighter-rouge">tbl_df</code>:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>as_data_frame(iris) 
tbl_df(iris) 
</code></pre></div></div>

<h3 id="reading-and-writing-tabular-files">Reading and writing tabular files</h3>

<p>While the base R read/write utilities can be used for <code class="language-plaintext highlighter-rouge">data.frames</code>,
best time performance with the least amount of typing is achieved with
the export/import functions from the <code class="language-plaintext highlighter-rouge">readr</code> package. For very large
files the <code class="language-plaintext highlighter-rouge">fread</code> function from the <code class="language-plaintext highlighter-rouge">data.table</code> package achieves the
best time performance.</p>

<h4 id="import-with-readr">Import with <code class="language-plaintext highlighter-rouge">readr</code></h4>

<p>Import functions provided by <code class="language-plaintext highlighter-rouge">readr</code> include:</p>

<ul>
  <li><code class="language-plaintext highlighter-rouge">read_csv()</code>: comma separated (CSV) files</li>
  <li><code class="language-plaintext highlighter-rouge">read_tsv()</code>: tab separated files</li>
  <li><code class="language-plaintext highlighter-rouge">read_delim()</code>: general delimited files</li>
  <li><code class="language-plaintext highlighter-rouge">read_fwf()</code>: fixed width files</li>
  <li><code class="language-plaintext highlighter-rouge">read_table()</code>: tabular files where colums are separated by
  white-space.</li>
  <li><code class="language-plaintext highlighter-rouge">read_log()</code>: web log files</li>
</ul>

<p>Create a sample tab delimited file for import</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>write_tsv(iris, "iris.txt") # Creates sample file
</code></pre></div></div>

<p>Import with <code class="language-plaintext highlighter-rouge">read_tsv</code></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>iris_df &lt;- read_tsv("iris.txt") # Import with read_tbv from readr package
iris_df

## # A tibble: 150 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
##  1          5.1         3.5          1.4         0.2 setosa 
##  2          4.9         3            1.4         0.2 setosa 
##  3          4.7         3.2          1.3         0.2 setosa 
##  4          4.6         3.1          1.5         0.2 setosa 
##  5          5           3.6          1.4         0.2 setosa 
##  6          5.4         3.9          1.7         0.4 setosa 
##  7          4.6         3.4          1.4         0.3 setosa 
##  8          5           3.4          1.5         0.2 setosa 
##  9          4.4         2.9          1.4         0.2 setosa 
## 10          4.9         3.1          1.5         0.1 setosa 
## # … with 140 more rows
</code></pre></div></div>

<p>To import Google Sheets directly into R, see
<a href="http://girke.bioinformatics.ucr.edu/GEN242/mydoc_Rbasics_10.html">here</a>.</p>

<h4 id="fast-table-import-with-fread">Fast table import with <code class="language-plaintext highlighter-rouge">fread</code></h4>

<p>The <code class="language-plaintext highlighter-rouge">fread</code> function from the <code class="language-plaintext highlighter-rouge">data.table</code> package provides the best
time performance for reading large tabular files into R.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library(data.table)
iris_df &lt;- as_data_frame(fread("iris.txt")) # Import with fread and conversion to tibble
iris_df

## # A tibble: 150 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
##  1          5.1         3.5          1.4         0.2 setosa 
##  2          4.9         3            1.4         0.2 setosa 
##  3          4.7         3.2          1.3         0.2 setosa 
##  4          4.6         3.1          1.5         0.2 setosa 
##  5          5           3.6          1.4         0.2 setosa 
##  6          5.4         3.9          1.7         0.4 setosa 
##  7          4.6         3.4          1.4         0.3 setosa 
##  8          5           3.4          1.5         0.2 setosa 
##  9          4.4         2.9          1.4         0.2 setosa 
## 10          4.9         3.1          1.5         0.1 setosa 
## # … with 140 more rows
</code></pre></div></div>

<p>Note: to ignore lines starting with comment signs, one can pass on to
<code class="language-plaintext highlighter-rouge">fread</code> a shell command for preprocessing the file. The following
example illustrates this option.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>fread("grep -v '^#' iris.txt") 
</code></pre></div></div>

<h4 id="export-with-readr">Export with <code class="language-plaintext highlighter-rouge">readr</code></h4>

<p>Export function provided by <code class="language-plaintext highlighter-rouge">readr</code> inlcude</p>

<ul>
  <li><code class="language-plaintext highlighter-rouge">write_delim()</code>: general delimited files</li>
  <li><code class="language-plaintext highlighter-rouge">write_csv()</code>: comma separated (CSV) files</li>
  <li><code class="language-plaintext highlighter-rouge">write_excel_csv()</code>: excel style CSV files</li>
  <li><code class="language-plaintext highlighter-rouge">write_tsv()</code>: tab separated files</li>
</ul>

<p>For instance, the <code class="language-plaintext highlighter-rouge">write_tsv</code> function writes a <code class="language-plaintext highlighter-rouge">data.frame</code> or <code class="language-plaintext highlighter-rouge">tibble</code>
to a tab delimited file with much nicer default settings than the base R
<code class="language-plaintext highlighter-rouge">write.table</code> function.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>write_tsv(iris_df, "iris.txt")
</code></pre></div></div>

<h3 id="column-and-row-binds">Column and row binds</h3>

<p>The equivalents to base R’s <code class="language-plaintext highlighter-rouge">rbind</code> and <code class="language-plaintext highlighter-rouge">cbind</code> are <code class="language-plaintext highlighter-rouge">bind_rows</code> and
<code class="language-plaintext highlighter-rouge">bind_cols</code>, respectively.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>bind_cols(iris_df, iris_df)

## # A tibble: 150 x 10
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species Sepal.Length1 Sepal.Width1
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;           &lt;dbl&gt;        &lt;dbl&gt;
##  1          5.1         3.5          1.4         0.2 setosa            5.1          3.5
##  2          4.9         3            1.4         0.2 setosa            4.9          3  
##  3          4.7         3.2          1.3         0.2 setosa            4.7          3.2
##  4          4.6         3.1          1.5         0.2 setosa            4.6          3.1
##  5          5           3.6          1.4         0.2 setosa            5            3.6
##  6          5.4         3.9          1.7         0.4 setosa            5.4          3.9
##  7          4.6         3.4          1.4         0.3 setosa            4.6          3.4
##  8          5           3.4          1.5         0.2 setosa            5            3.4
##  9          4.4         2.9          1.4         0.2 setosa            4.4          2.9
## 10          4.9         3.1          1.5         0.1 setosa            4.9          3.1
## # … with 140 more rows, and 3 more variables: Petal.Length1 &lt;dbl&gt;, Petal.Width1 &lt;dbl&gt;,
## #   Species1 &lt;chr&gt;

bind_rows(iris_df, iris_df)

## # A tibble: 300 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
##  1          5.1         3.5          1.4         0.2 setosa 
##  2          4.9         3            1.4         0.2 setosa 
##  3          4.7         3.2          1.3         0.2 setosa 
##  4          4.6         3.1          1.5         0.2 setosa 
##  5          5           3.6          1.4         0.2 setosa 
##  6          5.4         3.9          1.7         0.4 setosa 
##  7          4.6         3.4          1.4         0.3 setosa 
##  8          5           3.4          1.5         0.2 setosa 
##  9          4.4         2.9          1.4         0.2 setosa 
## 10          4.9         3.1          1.5         0.1 setosa 
## # … with 290 more rows
</code></pre></div></div>

<h3 id="extract-column-as-vector">Extract column as vector</h3>

<p>The subsetting operators <code class="language-plaintext highlighter-rouge">[[</code> and <code class="language-plaintext highlighter-rouge">$</code>can be used to extract from a
<code class="language-plaintext highlighter-rouge">tibble</code> single columns as vector.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>iris_df[[5]][1:12]

##  [1] "setosa" "setosa" "setosa" "setosa" "setosa" "setosa" "setosa" "setosa" "setosa" "setosa"
## [11] "setosa" "setosa"

iris_df$Species[1:12]

##  [1] "setosa" "setosa" "setosa" "setosa" "setosa" "setosa" "setosa" "setosa" "setosa" "setosa"
## [11] "setosa" "setosa"
</code></pre></div></div>

<h3 id="important-dplyr-functions">Important <code class="language-plaintext highlighter-rouge">dplyr</code> functions</h3>

<ol>
  <li><code class="language-plaintext highlighter-rouge">filter()</code> and <code class="language-plaintext highlighter-rouge">slice()</code></li>
  <li><code class="language-plaintext highlighter-rouge">arrange()</code></li>
  <li><code class="language-plaintext highlighter-rouge">select()</code> and <code class="language-plaintext highlighter-rouge">rename()</code></li>
  <li><code class="language-plaintext highlighter-rouge">distinct()</code></li>
  <li><code class="language-plaintext highlighter-rouge">mutate()</code> and <code class="language-plaintext highlighter-rouge">transmute()</code></li>
  <li><code class="language-plaintext highlighter-rouge">summarise()</code></li>
  <li><code class="language-plaintext highlighter-rouge">sample_n()</code> and <code class="language-plaintext highlighter-rouge">sample_frac()</code></li>
</ol>

<h3 id="slice-and-filter-functions">Slice and filter functions</h3>

<h4 id="filter-function">Filter function</h4>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>filter(iris_df, Sepal.Length &gt; 7.5, Species=="virginica")

## # A tibble: 6 x 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;    
## 1          7.6         3            6.6         2.1 virginica
## 2          7.7         3.8          6.7         2.2 virginica
## 3          7.7         2.6          6.9         2.3 virginica
## 4          7.7         2.8          6.7         2   virginica
## 5          7.9         3.8          6.4         2   virginica
## 6          7.7         3            6.1         2.3 virginica
</code></pre></div></div>

<h4 id="base-r-code-equivalent">Base R code equivalent</h4>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>iris_df[iris_df[, "Sepal.Length"] &gt; 7.5 &amp; iris_df[, "Species"]=="virginica", ]

## # A tibble: 6 x 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;    
## 1          7.6         3            6.6         2.1 virginica
## 2          7.7         3.8          6.7         2.2 virginica
## 3          7.7         2.6          6.9         2.3 virginica
## 4          7.7         2.8          6.7         2   virginica
## 5          7.9         3.8          6.4         2   virginica
## 6          7.7         3            6.1         2.3 virginica
</code></pre></div></div>

<h4 id="including-boolean-operators">Including boolean operators</h4>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>filter(iris_df, Sepal.Length &gt; 7.5 | Sepal.Length &lt; 5.5, Species=="virginica")

## # A tibble: 7 x 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;    
## 1          7.6         3            6.6         2.1 virginica
## 2          4.9         2.5          4.5         1.7 virginica
## 3          7.7         3.8          6.7         2.2 virginica
## 4          7.7         2.6          6.9         2.3 virginica
## 5          7.7         2.8          6.7         2   virginica
## 6          7.9         3.8          6.4         2   virginica
## 7          7.7         3            6.1         2.3 virginica
</code></pre></div></div>

<h4 id="subset-rows-by-position">Subset rows by position</h4>

<p><code class="language-plaintext highlighter-rouge">dplyr</code> approach</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>slice(iris_df, 1:2)

## # A tibble: 2 x 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
## 1          5.1         3.5          1.4         0.2 setosa 
## 2          4.9         3            1.4         0.2 setosa
</code></pre></div></div>

<p>Base R code equivalent</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>iris_df[1:2,]

## # A tibble: 2 x 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
## 1          5.1         3.5          1.4         0.2 setosa 
## 2          4.9         3            1.4         0.2 setosa
</code></pre></div></div>

<h4 id="subset-rows-by-names">Subset rows by names</h4>

<p>Since <code class="language-plaintext highlighter-rouge">tibbles</code> do not contain row names, row wise subsetting via the
<code class="language-plaintext highlighter-rouge">[,]</code> operator cannot be used. However, the corresponding behavior can
be achieved by passing to <code class="language-plaintext highlighter-rouge">select</code> a row position index obtained by
basic R intersect utilities such as <code class="language-plaintext highlighter-rouge">match</code>.</p>

<p>Create a suitable test <code class="language-plaintext highlighter-rouge">tibble</code></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>df1 &lt;- bind_cols(data_frame(ids1=paste0("g", 1:10)), as_data_frame(matrix(1:40, 10, 4, dimnames=list(1:10, paste0("CA", 1:4)))))
df1

## # A tibble: 10 x 5
##    ids1    CA1   CA2   CA3   CA4
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
##  1 g1        1    11    21    31
##  2 g2        2    12    22    32
##  3 g3        3    13    23    33
##  4 g4        4    14    24    34
##  5 g5        5    15    25    35
##  6 g6        6    16    26    36
##  7 g7        7    17    27    37
##  8 g8        8    18    28    38
##  9 g9        9    19    29    39
## 10 g10      10    20    30    40
</code></pre></div></div>

<p><code class="language-plaintext highlighter-rouge">dplyr</code> approach</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>slice(df1, match(c("g10", "g4", "g4"), df1$ids1))

## # A tibble: 3 x 5
##   ids1    CA1   CA2   CA3   CA4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1 g10      10    20    30    40
## 2 g4        4    14    24    34
## 3 g4        4    14    24    34
</code></pre></div></div>

<p>Base R equivalent</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>df1_old &lt;- as.data.frame(df1)
rownames(df1_old) &lt;- df1_old[,1]
df1_old[c("g10", "g4", "g4"),]

##      ids1 CA1 CA2 CA3 CA4
## g10   g10  10  20  30  40
## g4     g4   4  14  24  34
## g4.1   g4   4  14  24  34
</code></pre></div></div>

<h3 id="sorting-with-arrange">Sorting with <code class="language-plaintext highlighter-rouge">arrange</code></h3>

<p>Row-wise ordering based on specific columns</p>

<p><code class="language-plaintext highlighter-rouge">dplyr</code> approach</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>arrange(iris_df, Species, Sepal.Length, Sepal.Width)

## # A tibble: 150 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
##  1          4.3         3            1.1         0.1 setosa 
##  2          4.4         2.9          1.4         0.2 setosa 
##  3          4.4         3            1.3         0.2 setosa 
##  4          4.4         3.2          1.3         0.2 setosa 
##  5          4.5         2.3          1.3         0.3 setosa 
##  6          4.6         3.1          1.5         0.2 setosa 
##  7          4.6         3.2          1.4         0.2 setosa 
##  8          4.6         3.4          1.4         0.3 setosa 
##  9          4.6         3.6          1           0.2 setosa 
## 10          4.7         3.2          1.3         0.2 setosa 
## # … with 140 more rows
</code></pre></div></div>

<p>For ordering descendingly use <code class="language-plaintext highlighter-rouge">desc()</code> function</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>arrange(iris_df, desc(Species), Sepal.Length, Sepal.Width)

## # A tibble: 150 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;    
##  1          4.9         2.5          4.5         1.7 virginica
##  2          5.6         2.8          4.9         2   virginica
##  3          5.7         2.5          5           2   virginica
##  4          5.8         2.7          5.1         1.9 virginica
##  5          5.8         2.7          5.1         1.9 virginica
##  6          5.8         2.8          5.1         2.4 virginica
##  7          5.9         3            5.1         1.8 virginica
##  8          6           2.2          5           1.5 virginica
##  9          6           3            4.8         1.8 virginica
## 10          6.1         2.6          5.6         1.4 virginica
## # … with 140 more rows
</code></pre></div></div>

<p>Base R code equivalent</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>iris_df[order(iris_df$Species, iris_df$Sepal.Length, iris_df$Sepal.Width), ]

## # A tibble: 150 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;  
##  1          4.3         3            1.1         0.1 setosa 
##  2          4.4         2.9          1.4         0.2 setosa 
##  3          4.4         3            1.3         0.2 setosa 
##  4          4.4         3.2          1.3         0.2 setosa 
##  5          4.5         2.3          1.3         0.3 setosa 
##  6          4.6         3.1          1.5         0.2 setosa 
##  7          4.6         3.2          1.4         0.2 setosa 
##  8          4.6         3.4          1.4         0.3 setosa 
##  9          4.6         3.6          1           0.2 setosa 
## 10          4.7         3.2          1.3         0.2 setosa 
## # … with 140 more rows

iris_df[order(iris_df$Species, decreasing=TRUE), ] 

## # A tibble: 150 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species  
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;    
##  1          6.3         3.3          6           2.5 virginica
##  2          5.8         2.7          5.1         1.9 virginica
##  3          7.1         3            5.9         2.1 virginica
##  4          6.3         2.9          5.6         1.8 virginica
##  5          6.5         3            5.8         2.2 virginica
##  6          7.6         3            6.6         2.1 virginica
##  7          4.9         2.5          4.5         1.7 virginica
##  8          7.3         2.9          6.3         1.8 virginica
##  9          6.7         2.5          5.8         1.8 virginica
## 10          7.2         3.6          6.1         2.5 virginica
## # … with 140 more rows
</code></pre></div></div>

<h3 id="select-columns-with-select">Select columns with <code class="language-plaintext highlighter-rouge">select</code></h3>

<p>Select specific columns</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>select(iris_df, Species, Petal.Length, Sepal.Length)

## # A tibble: 150 x 3
##    Species Petal.Length Sepal.Length
##    &lt;chr&gt;          &lt;dbl&gt;        &lt;dbl&gt;
##  1 setosa           1.4          5.1
##  2 setosa           1.4          4.9
##  3 setosa           1.3          4.7
##  4 setosa           1.5          4.6
##  5 setosa           1.4          5  
##  6 setosa           1.7          5.4
##  7 setosa           1.4          4.6
##  8 setosa           1.5          5  
##  9 setosa           1.4          4.4
## 10 setosa           1.5          4.9
## # … with 140 more rows
</code></pre></div></div>

<p>Select range of columns by name</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>select(iris_df, Sepal.Length : Petal.Width)

## # A tibble: 150 x 4
##    Sepal.Length Sepal.Width Petal.Length Petal.Width
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
##  1          5.1         3.5          1.4         0.2
##  2          4.9         3            1.4         0.2
##  3          4.7         3.2          1.3         0.2
##  4          4.6         3.1          1.5         0.2
##  5          5           3.6          1.4         0.2
##  6          5.4         3.9          1.7         0.4
##  7          4.6         3.4          1.4         0.3
##  8          5           3.4          1.5         0.2
##  9          4.4         2.9          1.4         0.2
## 10          4.9         3.1          1.5         0.1
## # … with 140 more rows
</code></pre></div></div>

<p>Drop specific columns (here range)</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>select(iris_df, -(Sepal.Length : Petal.Width))

## # A tibble: 150 x 1
##    Species
##    &lt;chr&gt;  
##  1 setosa 
##  2 setosa 
##  3 setosa 
##  4 setosa 
##  5 setosa 
##  6 setosa 
##  7 setosa 
##  8 setosa 
##  9 setosa 
## 10 setosa 
## # … with 140 more rows
</code></pre></div></div>

<h3 id="renaming-columns-with-rename">Renaming columns with <code class="language-plaintext highlighter-rouge">rename</code></h3>

<p><code class="language-plaintext highlighter-rouge">dplyr</code> approach</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>rename(iris_df, new_col_name = Species)

## # A tibble: 150 x 5
##    Sepal.Length Sepal.Width Petal.Length Petal.Width new_col_name
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;       
##  1          5.1         3.5          1.4         0.2 setosa      
##  2          4.9         3            1.4         0.2 setosa      
##  3          4.7         3.2          1.3         0.2 setosa      
##  4          4.6         3.1          1.5         0.2 setosa      
##  5          5           3.6          1.4         0.2 setosa      
##  6          5.4         3.9          1.7         0.4 setosa      
##  7          4.6         3.4          1.4         0.3 setosa      
##  8          5           3.4          1.5         0.2 setosa      
##  9          4.4         2.9          1.4         0.2 setosa      
## 10          4.9         3.1          1.5         0.1 setosa      
## # … with 140 more rows
</code></pre></div></div>

<p>Base R code approach</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>colnames(iris_df)[colnames(iris_df)=="Species"] &lt;- "new_col_names"
</code></pre></div></div>

<h3 id="obtain-unique-rows-with-distinct">Obtain unique rows with <code class="language-plaintext highlighter-rouge">distinct</code></h3>

<p><code class="language-plaintext highlighter-rouge">dplyr</code> approach</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>distinct(iris_df, Species, .keep_all=TRUE)

## # A tibble: 3 x 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species   
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;     
## 1          5.1         3.5          1.4         0.2 setosa    
## 2          7           3.2          4.7         1.4 versicolor
## 3          6.3         3.3          6           2.5 virginica
</code></pre></div></div>

<p>Base R code approach</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>iris_df[!duplicated(iris_df$Species),]

## # A tibble: 3 x 5
##   Sepal.Length Sepal.Width Petal.Length Petal.Width Species   
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;     
## 1          5.1         3.5          1.4         0.2 setosa    
## 2          7           3.2          4.7         1.4 versicolor
## 3          6.3         3.3          6           2.5 virginica
</code></pre></div></div>

<h3 id="add-columns">Add columns</h3>

<h4 id="mutate"><code class="language-plaintext highlighter-rouge">mutate</code></h4>

<p>The <code class="language-plaintext highlighter-rouge">mutate</code> function allows to append columns to existing ones.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mutate(iris_df, Ratio = Sepal.Length / Sepal.Width, Sum = Sepal.Length + Sepal.Width)

## # A tibble: 150 x 7
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species Ratio   Sum
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;
##  1          5.1         3.5          1.4         0.2 setosa   1.46   8.6
##  2          4.9         3            1.4         0.2 setosa   1.63   7.9
##  3          4.7         3.2          1.3         0.2 setosa   1.47   7.9
##  4          4.6         3.1          1.5         0.2 setosa   1.48   7.7
##  5          5           3.6          1.4         0.2 setosa   1.39   8.6
##  6          5.4         3.9          1.7         0.4 setosa   1.38   9.3
##  7          4.6         3.4          1.4         0.3 setosa   1.35   8  
##  8          5           3.4          1.5         0.2 setosa   1.47   8.4
##  9          4.4         2.9          1.4         0.2 setosa   1.52   7.3
## 10          4.9         3.1          1.5         0.1 setosa   1.58   8  
## # … with 140 more rows
</code></pre></div></div>

<h4 id="transmute"><code class="language-plaintext highlighter-rouge">transmute</code></h4>

<p>The <code class="language-plaintext highlighter-rouge">transmute</code> function does the same as <code class="language-plaintext highlighter-rouge">mutate</code> but drops existing
columns</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>transmute(iris_df, Ratio = Sepal.Length / Sepal.Width, Sum = Sepal.Length + Sepal.Width)

## # A tibble: 150 x 2
##    Ratio   Sum
##    &lt;dbl&gt; &lt;dbl&gt;
##  1  1.46   8.6
##  2  1.63   7.9
##  3  1.47   7.9
##  4  1.48   7.7
##  5  1.39   8.6
##  6  1.38   9.3
##  7  1.35   8  
##  8  1.47   8.4
##  9  1.52   7.3
## 10  1.58   8  
## # … with 140 more rows
</code></pre></div></div>

<h4 id="bind_cols"><code class="language-plaintext highlighter-rouge">bind_cols</code></h4>

<p>The <code class="language-plaintext highlighter-rouge">bind_cols</code> function is the equivalent of <code class="language-plaintext highlighter-rouge">cbind</code> in base R. To add
rows, use the corresponding <code class="language-plaintext highlighter-rouge">bind_rows</code> function.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>bind_cols(iris_df, iris_df)

## # A tibble: 150 x 10
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species Sepal.Length1 Sepal.Width1
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;           &lt;dbl&gt;        &lt;dbl&gt;
##  1          5.1         3.5          1.4         0.2 setosa            5.1          3.5
##  2          4.9         3            1.4         0.2 setosa            4.9          3  
##  3          4.7         3.2          1.3         0.2 setosa            4.7          3.2
##  4          4.6         3.1          1.5         0.2 setosa            4.6          3.1
##  5          5           3.6          1.4         0.2 setosa            5            3.6
##  6          5.4         3.9          1.7         0.4 setosa            5.4          3.9
##  7          4.6         3.4          1.4         0.3 setosa            4.6          3.4
##  8          5           3.4          1.5         0.2 setosa            5            3.4
##  9          4.4         2.9          1.4         0.2 setosa            4.4          2.9
## 10          4.9         3.1          1.5         0.1 setosa            4.9          3.1
## # … with 140 more rows, and 3 more variables: Petal.Length1 &lt;dbl&gt;, Petal.Width1 &lt;dbl&gt;,
## #   Species1 &lt;chr&gt;
</code></pre></div></div>

<h3 id="summarize-data">Summarize data</h3>

<p>Summary calculation on single column</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>summarize(iris_df, mean(Petal.Length))

## # A tibble: 1 x 1
##   `mean(Petal.Length)`
##                  &lt;dbl&gt;
## 1                 3.76
</code></pre></div></div>

<p>Summary calculation on many columns</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>summarize_all(iris_df[,1:4], mean)

## # A tibble: 1 x 4
##   Sepal.Length Sepal.Width Petal.Length Petal.Width
##          &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
## 1         5.84        3.06         3.76        1.20
</code></pre></div></div>

<p>Summarize by grouping column</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>summarize(group_by(iris_df, Species), mean(Petal.Length))

## # A tibble: 3 x 2
##   Species    `mean(Petal.Length)`
##   &lt;chr&gt;                     &lt;dbl&gt;
## 1 setosa                     1.46
## 2 versicolor                 4.26
## 3 virginica                  5.55
</code></pre></div></div>

<p>Aggregate summaries</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>summarize_all(group_by(iris_df, Species), mean) 

## # A tibble: 3 x 5
##   Species    Sepal.Length Sepal.Width Petal.Length Petal.Width
##   &lt;chr&gt;             &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt;
## 1 setosa             5.01        3.43         1.46       0.246
## 2 versicolor         5.94        2.77         4.26       1.33 
## 3 virginica          6.59        2.97         5.55       2.03
</code></pre></div></div>

<p>Note: <code class="language-plaintext highlighter-rouge">group_by</code> does the looping for the user similar to <code class="language-plaintext highlighter-rouge">aggregate</code> or
<code class="language-plaintext highlighter-rouge">tapply</code>.</p>

<h3 id="merging-tibbles">Merging tibbles</h3>

<p>The <code class="language-plaintext highlighter-rouge">dplyr</code> package provides several join functions for merging
<code class="language-plaintext highlighter-rouge">tibbles</code> by a common key column similar to the <code class="language-plaintext highlighter-rouge">merge</code> function in base
R. These <code class="language-plaintext highlighter-rouge">*_join</code> functions include:</p>

<ul>
  <li><code class="language-plaintext highlighter-rouge">inner_join()</code>: returns join only for rows matching among both
  <code class="language-plaintext highlighter-rouge">tibbles</code></li>
  <li><code class="language-plaintext highlighter-rouge">full_join()</code>: returns join for all (matching and non-matching) rows
  of two <code class="language-plaintext highlighter-rouge">tibbles</code></li>
  <li><code class="language-plaintext highlighter-rouge">left_join()</code>: returns join for all rows in first <code class="language-plaintext highlighter-rouge">tibble</code></li>
  <li><code class="language-plaintext highlighter-rouge">right_join()</code>: returns join for all rows in second <code class="language-plaintext highlighter-rouge">tibble</code></li>
  <li><code class="language-plaintext highlighter-rouge">anti_join()</code>: returns for first <code class="language-plaintext highlighter-rouge">tibble</code> only those rows that have
  no match in the second one</li>
</ul>

<p>Sample <code class="language-plaintext highlighter-rouge">tibbles</code> to illustrate <code class="language-plaintext highlighter-rouge">*.join</code> functions.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>df1 &lt;- bind_cols(data_frame(ids1=paste0("g", 1:10)), as_data_frame(matrix(1:40, 10, 4, dimnames=list(1:10, paste0("CA", 1:4)))))
df1

## # A tibble: 10 x 5
##    ids1    CA1   CA2   CA3   CA4
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
##  1 g1        1    11    21    31
##  2 g2        2    12    22    32
##  3 g3        3    13    23    33
##  4 g4        4    14    24    34
##  5 g5        5    15    25    35
##  6 g6        6    16    26    36
##  7 g7        7    17    27    37
##  8 g8        8    18    28    38
##  9 g9        9    19    29    39
## 10 g10      10    20    30    40

df2 &lt;- bind_cols(data_frame(ids2=paste0("g", c(2,5,11,12))), as_data_frame(matrix(1:16, 4, 4, dimnames=list(1:4, paste0("CB", 1:4)))))
df2

## # A tibble: 4 x 5
##   ids2    CB1   CB2   CB3   CB4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1 g2        1     5     9    13
## 2 g5        2     6    10    14
## 3 g11       3     7    11    15
## 4 g12       4     8    12    16
</code></pre></div></div>

<h4 id="inner-join">Inner join</h4>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>inner_join(df1, df2, by=c("ids1"="ids2"))

## # A tibble: 2 x 9
##   ids1    CA1   CA2   CA3   CA4   CB1   CB2   CB3   CB4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1 g2        2    12    22    32     1     5     9    13
## 2 g5        5    15    25    35     2     6    10    14
</code></pre></div></div>

<h4 id="left-join">Left join</h4>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>left_join(df1, df2, by=c("ids1"="ids2"))

## # A tibble: 10 x 9
##    ids1    CA1   CA2   CA3   CA4   CB1   CB2   CB3   CB4
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
##  1 g1        1    11    21    31    NA    NA    NA    NA
##  2 g2        2    12    22    32     1     5     9    13
##  3 g3        3    13    23    33    NA    NA    NA    NA
##  4 g4        4    14    24    34    NA    NA    NA    NA
##  5 g5        5    15    25    35     2     6    10    14
##  6 g6        6    16    26    36    NA    NA    NA    NA
##  7 g7        7    17    27    37    NA    NA    NA    NA
##  8 g8        8    18    28    38    NA    NA    NA    NA
##  9 g9        9    19    29    39    NA    NA    NA    NA
## 10 g10      10    20    30    40    NA    NA    NA    NA
</code></pre></div></div>

<h4 id="right-join">Right join</h4>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>right_join(df1, df2, by=c("ids1"="ids2"))

## # A tibble: 4 x 9
##   ids1    CA1   CA2   CA3   CA4   CB1   CB2   CB3   CB4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1 g2        2    12    22    32     1     5     9    13
## 2 g5        5    15    25    35     2     6    10    14
## 3 g11      NA    NA    NA    NA     3     7    11    15
## 4 g12      NA    NA    NA    NA     4     8    12    16
</code></pre></div></div>

<h4 id="full-join">Full join</h4>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>full_join(df1, df2, by=c("ids1"="ids2"))

## # A tibble: 12 x 9
##    ids1    CA1   CA2   CA3   CA4   CB1   CB2   CB3   CB4
##    &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
##  1 g1        1    11    21    31    NA    NA    NA    NA
##  2 g2        2    12    22    32     1     5     9    13
##  3 g3        3    13    23    33    NA    NA    NA    NA
##  4 g4        4    14    24    34    NA    NA    NA    NA
##  5 g5        5    15    25    35     2     6    10    14
##  6 g6        6    16    26    36    NA    NA    NA    NA
##  7 g7        7    17    27    37    NA    NA    NA    NA
##  8 g8        8    18    28    38    NA    NA    NA    NA
##  9 g9        9    19    29    39    NA    NA    NA    NA
## 10 g10      10    20    30    40    NA    NA    NA    NA
## 11 g11      NA    NA    NA    NA     3     7    11    15
## 12 g12      NA    NA    NA    NA     4     8    12    16
</code></pre></div></div>

<h4 id="anti-join">Anti join</h4>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>anti_join(df1, df2, by=c("ids1"="ids2"))

## # A tibble: 8 x 5
##   ids1    CA1   CA2   CA3   CA4
##   &lt;chr&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt; &lt;int&gt;
## 1 g1        1    11    21    31
## 2 g3        3    13    23    33
## 3 g4        4    14    24    34
## 4 g6        6    16    26    36
## 5 g7        7    17    27    37
## 6 g8        8    18    28    38
## 7 g9        9    19    29    39
## 8 g10      10    20    30    40
</code></pre></div></div>

<p>For additional join options users want to cosult the <code class="language-plaintext highlighter-rouge">*_join</code> help
pages.</p>

<h3 id="chaining">Chaining</h3>

<p>To simplify chaining of serveral operations, <code class="language-plaintext highlighter-rouge">dplyr</code> provides the <code class="language-plaintext highlighter-rouge">%&gt;%</code>
operator, where <code class="language-plaintext highlighter-rouge">x %&gt;% f(y)</code> turns into <code class="language-plaintext highlighter-rouge">f(x, y)</code>. This way one can pipe
together multiple operations by writing them from left-to-right or
top-to-bottom. This makes for easy to type and readable code.</p>

<h4 id="example-1">Example 1</h4>

<p>Series of data manipulations and export</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>read_tsv("iris.txt") %&gt;% # Import with read_tbv from readr package
    as_tibble() %&gt;% # Declare tibble to use
    select(Sepal.Length:Species) %&gt;% # Select columns
    filter(Species=="setosa") %&gt;% # Filter rows by some value
    arrange(Sepal.Length) %&gt;% # Sort by some column
    mutate(Subtract=Petal.Length - Petal.Width) # Calculate and append

## # A tibble: 50 x 6
##    Sepal.Length Sepal.Width Petal.Length Petal.Width Species Subtract
##           &lt;dbl&gt;       &lt;dbl&gt;        &lt;dbl&gt;       &lt;dbl&gt; &lt;chr&gt;      &lt;dbl&gt;
##  1          4.3         3            1.1         0.1 setosa      1   
##  2          4.4         2.9          1.4         0.2 setosa      1.2 
##  3          4.4         3            1.3         0.2 setosa      1.1 
##  4          4.4         3.2          1.3         0.2 setosa      1.1 
##  5          4.5         2.3          1.3         0.3 setosa      1   
##  6          4.6         3.1          1.5         0.2 setosa      1.3 
##  7          4.6         3.4          1.4         0.3 setosa      1.10
##  8          4.6         3.6          1           0.2 setosa      0.8 
##  9          4.6         3.2          1.4         0.2 setosa      1.2 
## 10          4.7         3.2          1.3         0.2 setosa      1.1 
## # … with 40 more rows

    # write_tsv("iris.txt") # Export to file, omitted here to show result 
</code></pre></div></div>

<h4 id="example-2">Example 2</h4>

<p>Series of summary calculations for grouped data (<code class="language-plaintext highlighter-rouge">group_by</code>)</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>iris_df %&gt;% # Declare tibble to use 
    group_by(Species) %&gt;% # Group by species
    summarize(Mean_Sepal.Length=mean(Sepal.Length), 
              Max_Sepal.Length=max(Sepal.Length),
              Min_Sepal.Length=min(Sepal.Length),
              SD_Sepal.Length=sd(Sepal.Length),
              Total=n()) 

## # A tibble: 3 x 6
##   Species    Mean_Sepal.Length Max_Sepal.Length Min_Sepal.Length SD_Sepal.Length Total
##   &lt;chr&gt;                  &lt;dbl&gt;            &lt;dbl&gt;            &lt;dbl&gt;           &lt;dbl&gt; &lt;int&gt;
## 1 setosa                  5.01              5.8              4.3           0.352    50
## 2 versicolor              5.94              7                4.9           0.516    50
## 3 virginica               6.59              7.9              4.9           0.636    50
</code></pre></div></div>

<h4 id="example-3">Example 3</h4>

<p>Combining <code class="language-plaintext highlighter-rouge">dplyr</code> chaining with <code class="language-plaintext highlighter-rouge">ggplot</code></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>iris_df %&gt;% 
    group_by(Species) %&gt;% 
    summarize_all(mean) %&gt;% 
    reshape2::melt(id.vars=c("Species"), variable.name = "Samples", value.name="Values") %&gt;%
    ggplot(aes(Samples, Values, fill = Species)) + 
        geom_bar(position="dodge", stat="identity")
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/plyr_chaining3-1.svg" alt="" /></p>

<h2 id="sqlite-databases">SQLite Databases</h2>

<p><code class="language-plaintext highlighter-rouge">SQLite</code> is a lightweight relational database solution. The <code class="language-plaintext highlighter-rouge">RSQLite</code>
package provides an easy to use interface to create, manage and query
<code class="language-plaintext highlighter-rouge">SQLite</code> databases directly from R. Basic instructions for using
<code class="language-plaintext highlighter-rouge">SQLite</code> from the command-line are available
<a href="https://www.sqlite.org/cli.html">here</a>. A short introduction to
<code class="language-plaintext highlighter-rouge">RSQLite</code> is available
<a href="https://github.com/rstats-db/RSQLite/blob/master/vignettes/RSQLite.Rmd">here</a>.</p>

<h3 id="loading-data-into-sqlite-databases">Loading data into SQLite databases</h3>

<p>The following loads two <code class="language-plaintext highlighter-rouge">data.frames</code> derived from the <code class="language-plaintext highlighter-rouge">iris</code> data set
(here <code class="language-plaintext highlighter-rouge">mydf1</code> and <code class="language-plaintext highlighter-rouge">mydf2</code>) into an SQLite database (here <code class="language-plaintext highlighter-rouge">test.db</code>).</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library(RSQLite)
unlink("test.db") # Delete any existing test.db
mydb &lt;- dbConnect(SQLite(), "test.db") # Creates database file test.db
mydf1 &lt;- data.frame(ids=paste0("id", seq_along(iris[,1])), iris)
mydf2 &lt;- mydf1[sample(seq_along(mydf1[,1]), 10),]
dbWriteTable(mydb, "mydf1", mydf1)
dbWriteTable(mydb, "mydf2", mydf2)
</code></pre></div></div>

<h3 id="list-names-of-tables-in-database">List names of tables in database</h3>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>dbListTables(mydb)

## [1] "mydf1" "mydf2"
</code></pre></div></div>

<h3 id="import-table-into-dataframe">Import table into <code class="language-plaintext highlighter-rouge">data.frame</code></h3>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>dbGetQuery(mydb, 'SELECT * FROM mydf2')

##      ids Sepal.Length Sepal.Width Petal.Length Petal.Width    Species
## 1   id60          5.2         2.7          3.9         1.4 versicolor
## 2   id71          5.9         3.2          4.8         1.8 versicolor
## 3   id87          6.7         3.1          4.7         1.5 versicolor
## 4  id136          7.7         3.0          6.1         2.3  virginica
## 5  id145          6.7         3.3          5.7         2.5  virginica
## 6   id59          6.6         2.9          4.6         1.3 versicolor
## 7  id103          7.1         3.0          5.9         2.1  virginica
## 8   id41          5.0         3.5          1.3         0.3     setosa
## 9  id150          5.9         3.0          5.1         1.8  virginica
## 10  id63          6.0         2.2          4.0         1.0 versicolor
</code></pre></div></div>

<h3 id="query-database">Query database</h3>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>dbGetQuery(mydb, 'SELECT * FROM mydf1 WHERE "Sepal.Length" &lt; 4.6')

##    ids Sepal.Length Sepal.Width Petal.Length Petal.Width Species
## 1  id9          4.4         2.9          1.4         0.2  setosa
## 2 id14          4.3         3.0          1.1         0.1  setosa
## 3 id39          4.4         3.0          1.3         0.2  setosa
## 4 id42          4.5         2.3          1.3         0.3  setosa
## 5 id43          4.4         3.2          1.3         0.2  setosa
</code></pre></div></div>

<h3 id="join-tables">Join tables</h3>

<p>The two tables can be joined on the shared <code class="language-plaintext highlighter-rouge">ids</code> column as follows.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>dbGetQuery(mydb, 'SELECT * FROM mydf1, mydf2 WHERE mydf1.ids = mydf2.ids')

##      ids Sepal.Length Sepal.Width Petal.Length Petal.Width    Species   ids Sepal.Length
## 1   id41          5.0         3.5          1.3         0.3     setosa  id41          5.0
## 2   id59          6.6         2.9          4.6         1.3 versicolor  id59          6.6
## 3   id60          5.2         2.7          3.9         1.4 versicolor  id60          5.2
## 4   id63          6.0         2.2          4.0         1.0 versicolor  id63          6.0
## 5   id71          5.9         3.2          4.8         1.8 versicolor  id71          5.9
## 6   id87          6.7         3.1          4.7         1.5 versicolor  id87          6.7
## 7  id103          7.1         3.0          5.9         2.1  virginica id103          7.1
## 8  id136          7.7         3.0          6.1         2.3  virginica id136          7.7
## 9  id145          6.7         3.3          5.7         2.5  virginica id145          6.7
## 10 id150          5.9         3.0          5.1         1.8  virginica id150          5.9
##    Sepal.Width Petal.Length Petal.Width    Species
## 1          3.5          1.3         0.3     setosa
## 2          2.9          4.6         1.3 versicolor
## 3          2.7          3.9         1.4 versicolor
## 4          2.2          4.0         1.0 versicolor
## 5          3.2          4.8         1.8 versicolor
## 6          3.1          4.7         1.5 versicolor
## 7          3.0          5.9         2.1  virginica
## 8          3.0          6.1         2.3  virginica
## 9          3.3          5.7         2.5  virginica
## 10         3.0          5.1         1.8  virginica
</code></pre></div></div>

<h2 id="graphics-in-r">Graphics in R</h2>

<h3 id="advantages">Advantages</h3>

<ul>
  <li>Powerful environment for visualizing scientific data</li>
  <li>Integrated graphics and statistics infrastructure</li>
  <li>Publication quality graphics</li>
  <li>Fully programmable</li>
  <li>Highly reproducible</li>
  <li>Full <a href="http://www.latex-project.org/">LaTeX</a> and Markdown support via
  <code class="language-plaintext highlighter-rouge">knitr</code> and <code class="language-plaintext highlighter-rouge">R markdown</code></li>
  <li>Vast number of R packages with graphics utilities</li>
</ul>

<h3 id="documentation-for-r-graphics">Documentation for R Graphics</h3>

<p><strong>General</strong></p>

<ul>
  <li>Graphics Task Page -
  <a href="http://cran.r-project.org/web/views/Graphics.html">URL</a></li>
  <li>R Graph Gallery -
  <a href="http://addictedtor.free.fr/graphiques/allgraph.php">URL</a></li>
  <li>R Graphical Manual -
  <a href="http://cged.genes.nig.ac.jp/RGM2/index.php">URL</a></li>
  <li>Paul Murrell’s book R (Grid) Graphics -
  <a href="http://www.stat.auckland.ac.nz/~paul/RGraphics/rgraphics.html">URL</a></li>
</ul>

<p><strong>Interactive graphics</strong></p>

<ul>
  <li>rggobi` (GGobi) - <a href="http://www.ggobi.org/">URL</a></li>
  <li><code class="language-plaintext highlighter-rouge">iplots</code> - <a href="http://www.rosuda.org/iplots/">URL</a></li>
  <li>Open GL (<code class="language-plaintext highlighter-rouge">rgl</code>) - <a href="http://rgl.neoscientists.org/gallery.shtml">URL</a></li>
</ul>

<h3 id="graphics-environments">Graphics Environments</h3>

<p><strong>Viewing and saving graphics in R</strong></p>

<ul>
  <li>On-screen graphics</li>
  <li>postscript, pdf, svg</li>
  <li>jpeg, png, wmf, tiff, …</li>
</ul>

<p><strong>Four major graphic environments</strong></p>

<ol>
  <li>Low-level infrastructure</li>
</ol>

<ul>
  <li>R Base Graphics (low- and high-level)</li>
  <li><code class="language-plaintext highlighter-rouge">grid</code>:
  <a href="http://www.stat.auckland.ac.nz/~paul/grid/grid.html">Manual</a></li>
</ul>

<ol>
  <li>High-level infrastructure \begin{itemize}</li>
</ol>

<ul>
  <li><code class="language-plaintext highlighter-rouge">lattice</code>: <a href="http://lmdvr.r-forge.r-project.org">Manual</a>,
  <a href="http://www.his.sunderland.ac.uk/~cs0her/Statistics/UsingLatticeGraphicsInR.htm">Intro</a>,
  <a href="http://www.amazon.com/Lattice-Multivariate-Data-Visualization-Use/dp/0387759689">Book</a></li>
  <li><code class="language-plaintext highlighter-rouge">ggplot2</code>: <a href="http://had.co.nz/ggplot2/">Manual</a>,
  <a href="http://www.ling.upenn.edu/~joseff/rstudy/summer2010_ggplot2_intro.html">Intro</a>,
  <a href="http://had.co.nz/ggplot2/book/">Book</a></li>
</ul>

<h3 id="base-graphics-overview">Base Graphics: Overview</h3>

<p><strong>Important high-level plotting functions</strong></p>

<ul>
  <li><code class="language-plaintext highlighter-rouge">plot</code>: generic x-y plotting</li>
  <li><code class="language-plaintext highlighter-rouge">barplot</code>: bar plots</li>
  <li><code class="language-plaintext highlighter-rouge">boxplot</code>: box-and-whisker plot</li>
  <li><code class="language-plaintext highlighter-rouge">hist</code>: histograms</li>
  <li><code class="language-plaintext highlighter-rouge">pie</code>: pie charts</li>
  <li><code class="language-plaintext highlighter-rouge">dotchart</code>: cleveland dot plots</li>
  <li><code class="language-plaintext highlighter-rouge">image, heatmap, contour, persp</code>: functions to generate image-like
  plots</li>
  <li><code class="language-plaintext highlighter-rouge">qqnorm, qqline, qqplot</code>: distribution comparison plots</li>
  <li><code class="language-plaintext highlighter-rouge">pairs, coplot</code>: display of multivariant data</li>
</ul>

<p><strong>Help on graphics functions</strong></p>

<ul>
  <li><code class="language-plaintext highlighter-rouge">?myfct</code></li>
  <li><code class="language-plaintext highlighter-rouge">?plot</code></li>
  <li><code class="language-plaintext highlighter-rouge">?par</code></li>
</ul>

<h4 id="preferred-object-types">Preferred Object Types</h4>

<ul>
  <li>Matrices and data frames</li>
  <li>Vectors</li>
  <li>Named vectors</li>
</ul>

<h3 id="scatter-plots">Scatter Plots</h3>

<h4 id="basic-scatter-plot">Basic Scatter Plot</h4>

<p>Sample data set for subsequent plots</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>set.seed(1410)
y &lt;- matrix(runif(30), ncol=3, dimnames=list(letters[1:10], LETTERS[1:3]))
</code></pre></div></div>

<p>Plot data</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>plot(y[,1], y[,2]) 
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/basic_scatter_plot-1.svg" alt="" /></p>

<h4 id="all-pairs">All pairs</h4>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pairs(y) 
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/pairs_scatter_plot-1.svg" alt="" /></p>

<h4 id="with-labels">With labels</h4>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>plot(y[,1], y[,2], pch=20, col="red", main="Symbols and Labels")
text(y[,1]+0.03, y[,2], rownames(y))
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/labels_scatter_plot-1.svg" alt="" /></p>

<h3 id="more-examples">More examples</h3>

<p><strong>Print instead of symbols the row names</strong></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>plot(y[,1], y[,2], type="n", main="Plot of Labels")
text(y[,1], y[,2], rownames(y)) 
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/row_scatter_plot-1.svg" alt="" /></p>

<p><strong>Usage of important plotting parameters</strong></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>grid(5, 5, lwd = 2) 
op &lt;- par(mar=c(8,8,8,8), bg="lightblue")
plot(y[,1], y[,2], type="p", col="red", cex.lab=1.2, cex.axis=1.2, 
     cex.main=1.2, cex.sub=1, lwd=4, pch=20, xlab="x label", 
     ylab="y label", main="My Main", sub="My Sub")
par(op)
</code></pre></div></div>

<p>__Important arguments_</p>

<ul>
  <li><code class="language-plaintext highlighter-rouge">mar</code>: specifies the margin sizes around the plotting area in order:
  <code class="language-plaintext highlighter-rouge">c(bottom, left, top, right)</code></li>
  <li><code class="language-plaintext highlighter-rouge">col</code>: color of symbols</li>
  <li><code class="language-plaintext highlighter-rouge">pch</code>: type of symbols, samples: <code class="language-plaintext highlighter-rouge">example(points)</code></li>
  <li><code class="language-plaintext highlighter-rouge">lwd</code>: size of symbols</li>
  <li><code class="language-plaintext highlighter-rouge">cex.*</code>: control font sizes</li>
  <li>For details see <code class="language-plaintext highlighter-rouge">?par</code></li>
</ul>

<h4 id="add-regression-line">Add regression line</h4>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>plot(y[,1], y[,2])
myline &lt;- lm(y[,2]~y[,1]); abline(myline, lwd=2) 
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/plot_regression-1.svg" alt="" /></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>summary(myline) 

## 
## Call:
## lm(formula = y[, 2] ~ y[, 1])
## 
## Residuals:
##      Min       1Q   Median       3Q      Max 
## -0.40357 -0.17912 -0.04299  0.22147  0.46623 
## 
## Coefficients:
##             Estimate Std. Error t value Pr(&gt;|t|)  
## (Intercept)   0.5764     0.2110   2.732   0.0258 *
## y[, 1]       -0.3647     0.3959  -0.921   0.3839  
## ---
## Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
## 
## Residual standard error: 0.3095 on 8 degrees of freedom
## Multiple R-squared:  0.09589,    Adjusted R-squared:  -0.01712 
## F-statistic: 0.8485 on 1 and 8 DF,  p-value: 0.3839
</code></pre></div></div>

<h4 id="log-scale">Log scale</h4>

<p>Same plot as above, but on log scale</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>plot(y[,1], y[,2], log="xy") 
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/plot_regression_log-1.svg" alt="" /></p>

<h4 id="add-a-mathematical-expression">Add a mathematical expression</h4>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>plot(y[,1], y[,2]); text(y[1,1], y[1,2], expression(sum(frac(1,sqrt(x^2*pi)))), cex=1.3) 
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/plot_regression_math-1.svg" alt="" /></p>

<h3 id="homework-3b">Homework 3B</h3>

<p>Homework 3B: <a href="http://girke.bioinformatics.ucr.edu/GEN242/mydoc_homework_03.html">Scatter
Plots</a></p>

<h3 id="line-plots">Line Plots</h3>

<h4 id="single-data-set">Single data set</h4>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>plot(y[,1], type="l", lwd=2, col="blue") 
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/plot_line_single-1.svg" alt="" /></p>

<h4 id="many-data-sets">Many Data Sets</h4>

<p>Plots line graph for all columns in data frame <code class="language-plaintext highlighter-rouge">y</code>. The <code class="language-plaintext highlighter-rouge">split.screen</code>
function is used in this example in a for loop to overlay several line
graphs in the same plot.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>split.screen(c(1,1)) 

## [1] 1

plot(y[,1], ylim=c(0,1), xlab="Measurement", ylab="Intensity", type="l", lwd=2, col=1)
for(i in 2:length(y[1,])) { 
    screen(1, new=FALSE)
    plot(y[,i], ylim=c(0,1), type="l", lwd=2, col=i, xaxt="n", yaxt="n", ylab="", xlab="", main="", bty="n") 
}
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/plot_line_many-1.svg" alt="" /></p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>close.screen(all=TRUE) 
</code></pre></div></div>

<h3 id="bar-plots">Bar Plots</h3>

<h4 id="basics">Basics</h4>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>barplot(y[1:4,], ylim=c(0, max(y[1:4,])+0.3), beside=TRUE, legend=letters[1:4]) 
text(labels=round(as.vector(as.matrix(y[1:4,])),2), x=seq(1.5, 13, by=1) + sort(rep(c(0,1,2), 4)), y=as.vector(as.matrix(y[1:4,]))+0.04) 
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/plot_bar_simple-1.svg" alt="" /></p>

<h4 id="error-bars">Error Bars</h4>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>bar &lt;- barplot(m &lt;- rowMeans(y) * 10, ylim=c(0, 10))
stdev &lt;- sd(t(y))
arrows(bar, m, bar, m + stdev, length=0.15, angle = 90)
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/plot_bar_error-1.svg" alt="" /></p>

<h3 id="histograms">Histograms</h3>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>hist(y, freq=TRUE, breaks=10)
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/plot_hist-1.svg" alt="" /></p>

<h3 id="density-plots">Density Plots</h3>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>plot(density(y), col="red")
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/plot_dens-1.svg" alt="" /></p>

<h3 id="pie-charts">Pie Charts</h3>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pie(y[,1], col=rainbow(length(y[,1]), start=0.1, end=0.8), clockwise=TRUE)
legend("topright", legend=row.names(y), cex=1.3, bty="n", pch=15, pt.cex=1.8, 
col=rainbow(length(y[,1]), start=0.1, end=0.8), ncol=1) 
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/plot_pie-1.svg" alt="" /></p>

<h3 id="color-selection-utilities">Color Selection Utilities</h3>

<p>Default color palette and how to change it</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>palette()

## [1] "black"   "#DF536B" "#61D04F" "#2297E6" "#28E2E5" "#CD0BBC" "#F5C710" "gray62"

palette(rainbow(5, start=0.1, end=0.2))
palette()

## [1] "#FF9900" "#FFBF00" "#FFE600" "#F2FF00" "#CCFF00"

palette("default")
</code></pre></div></div>

<p>The <code class="language-plaintext highlighter-rouge">gray</code> function allows to select any type of gray shades by
providing values from 0 to 1</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>gray(seq(0.1, 1, by= 0.2))

## [1] "#1A1A1A" "#4D4D4D" "#808080" "#B3B3B3" "#E6E6E6"
</code></pre></div></div>

<p>Color gradients with <code class="language-plaintext highlighter-rouge">colorpanel</code> function from <code class="language-plaintext highlighter-rouge">gplots</code> library`</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library(gplots)
colorpanel(5, "darkblue", "yellow", "white")

## [1] "#00008B" "#808046" "#FFFF00" "#FFFF80" "#FFFFFF"
</code></pre></div></div>

<p>Much more on colors in R see Earl Glynn’s color chart
<a href="http://research.stowers-institute.org/efg/R/Color/Chart/">here</a></p>

<h3 id="saving-graphics-to-file">Saving Graphics to File</h3>

<p>After the <code class="language-plaintext highlighter-rouge">pdf()</code> command all graphs are redirected to file <code class="language-plaintext highlighter-rouge">test.pdf</code>.
Works for all common formats similarly: jpeg, png, ps, tiff, …</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>pdf("test.pdf")
plot(1:10, 1:10)
dev.off() 
</code></pre></div></div>

<p>Generates Scalable Vector Graphics (SVG) files that can be edited in
vector graphics programs, such as InkScape.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library("RSvgDevice")
devSVG("test.svg")
plot(1:10, 1:10)
dev.off() 
</code></pre></div></div>

<h3 id="homework-3c">Homework 3C</h3>

<p>Homework 3C: <a href="http://girke.bioinformatics.ucr.edu/GEN242/mydoc_homework_03.html">Bar
Plots</a></p>

<h2 id="analysis-routine">Analysis Routine</h2>

<h3 id="overview">Overview</h3>

<p>The following exercise introduces a variety of useful data analysis
utilities in R.</p>

<h3 id="analysis-routine-data-import">Analysis Routine: Data Import</h3>

<ul>
  <li>
    <p><strong>Step 1</strong>: To get started with this exercise, direct your R session
  to a dedicated workshop directory and download into this directory
  the following sample tables. Then import the files into Excel and
  save them as tab delimited text files.</p>

    <ul>
      <li><a href="http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/MolecularWeight_tair7.xls">MolecularWeight_tair7.xls</a></li>
      <li><a href="http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/TargetP_analysis_tair7.xls">TargetP_analysis_tair7.xls</a></li>
    </ul>
  </li>
</ul>

<p><strong>Import the tables into R</strong></p>

<p>Import molecular weight table</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>my_mw &lt;- read.delim(file="MolecularWeight_tair7.xls", header=T, sep="\t") 
my_mw[1:2,]
</code></pre></div></div>

<p>Import subcelluar targeting table</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>my_target &lt;- read.delim(file="TargetP_analysis_tair7.xls", header=T, sep="\t") 
my_target[1:2,]
</code></pre></div></div>

<p>Online import of molecular weight table</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>my_mw &lt;- read.delim(file="http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/MolecularWeight_tair7.xls", header=T, sep="\t") 
my_mw[1:2,]

##   Sequence.id Molecular.Weight.Da. Residues
## 1 AT1G08520.1                83285      760
## 2 AT1G08530.1                27015      257
</code></pre></div></div>

<p>Online import of subcelluar targeting table</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>my_target &lt;- read.delim(file="http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/Samples/TargetP_analysis_tair7.xls", header=T, sep="\t") 
my_target[1:2,]

##      GeneName Loc   cTP   mTP    SP other
## 1 AT1G08520.1   C 0.822 0.137 0.029 0.039
## 2 AT1G08530.1   C 0.817 0.058 0.010 0.100
</code></pre></div></div>

<h3 id="merging-data-frames">Merging Data Frames</h3>

<ul>
  <li><strong>Step 2</strong>: Assign uniform gene ID column titles</li>
</ul>

<!-- -->

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>colnames(my_target)[1] &lt;- "ID"
colnames(my_mw)[1] &lt;- "ID" 
</code></pre></div></div>

<ul>
  <li><strong>Step 3</strong>: Merge the two tables based on common ID field</li>
</ul>

<!-- -->

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>my_mw_target &lt;- merge(my_mw, my_target, by.x="ID", by.y="ID", all.x=T)
</code></pre></div></div>

<ul>
  <li><strong>Step 4</strong>: Shorten one table before the merge and then remove the
  non-matching rows (NAs) in the merged file</li>
</ul>

<!-- -->

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>my_mw_target2a &lt;- merge(my_mw, my_target[1:40,], by.x="ID", by.y="ID", all.x=T)  # To remove non-matching rows, use the argument setting 'all=F'.
my_mw_target2 &lt;- na.omit(my_mw_target2a) # Removes rows containing "NAs" (non-matching rows).
</code></pre></div></div>

<ul>
  <li><strong>Homework 3D</strong>: How can the merge function in the previous step be
  executed so that only the common rows among the two data frames are
  returned? Prove that both methods - the two step version with
  <code class="language-plaintext highlighter-rouge">na.omit</code> and your method - return identical results.</li>
  <li><strong>Homework 3E</strong>: Replace all <code class="language-plaintext highlighter-rouge">NAs</code> in the data frame
  <code class="language-plaintext highlighter-rouge">my_mw_target2a</code> with zeros.</li>
</ul>

<h3 id="filtering-data">Filtering Data</h3>

<ul>
  <li><strong>Step 5</strong>: Retrieve all records with a value of greater than
  100,000 in ‘MW’ column and ‘C’ value in ‘Loc’ column (targeted to
  chloroplast).</li>
</ul>

<!-- -->

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>query &lt;- my_mw_target[my_mw_target[, 2] &gt; 100000 &amp; my_mw_target[, 4] == "C", ] 
query[1:4, ]

##              ID Molecular.Weight.Da. Residues Loc   cTP   mTP    SP other
## 219 AT1G02730.1               132588     1181   C 0.972 0.038 0.008 0.045
## 243 AT1G02890.1               136825     1252   C 0.748 0.529 0.011 0.013
## 281 AT1G03160.1               100732      912   C 0.871 0.235 0.011 0.007
## 547 AT1G05380.1               126360     1138   C 0.740 0.099 0.016 0.358

dim(query)

## [1] 170   8
</code></pre></div></div>

<ul>
  <li><strong>Homework 3F</strong>: How many protein entries in the <code class="language-plaintext highlighter-rouge">my</code>_mw<code class="language-plaintext highlighter-rouge">_target</code>
  data frame have a MW of greater then 4,000 and less then 5,000.
  Subset the data frame accordingly and sort it by MW to check that
  your result is correct.</li>
</ul>

<h3 id="string-substitutions">String Substitutions</h3>

<ul>
  <li><strong>Step 6</strong>: Use a regular expression in a substitute function to
  generate a separate ID column that lacks the gene model extensions.
  &lt;&lt;label=Exercise 4.7, eval=TRUE, echo=TRUE,
  keep.source=TRUE&gt;&gt;=</li>
</ul>

<!-- -->

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>my_mw_target3 &lt;- data.frame(loci=gsub("\\..*", "", as.character(my_mw_target[,1]), perl = TRUE), my_mw_target)
my_mw_target3[1:3,1:8]

##        loci          ID Molecular.Weight.Da. Residues Loc  cTP   mTP    SP
## 1 AT1G01010 AT1G01010.1                49426      429   _ 0.10 0.090 0.075
## 2 AT1G01020 AT1G01020.1                28092      245   * 0.01 0.636 0.158
## 3 AT1G01020 AT1G01020.2                21711      191   * 0.01 0.636 0.158
</code></pre></div></div>

<ul>
  <li><strong>Homework 3G</strong>: Retrieve those rows in <code class="language-plaintext highlighter-rouge">my_mw_target3</code> where the
  second column contains the following identifiers:
  <code class="language-plaintext highlighter-rouge">c("AT5G52930.1", "AT4G18950.1", "AT1G15385.1", "AT4G36500.1", "AT1G67530.1")</code>.
  Use the <code class="language-plaintext highlighter-rouge">%in%</code> function for this query. As an alternative approach,
  assign the second column to the row index of the data frame and then
  perform the same query again using the row index. Explain the
  difference of the two methods.</li>
</ul>

<h3 id="calculations-on-data-frames">Calculations on Data Frames</h3>

<ul>
  <li><strong>Step 7</strong>: Count the number of duplicates in the loci column with
  the <code class="language-plaintext highlighter-rouge">table</code> function and append the result to the data frame with
  the <code class="language-plaintext highlighter-rouge">cbind</code> function.</li>
</ul>

<!-- -->

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mycounts &lt;- table(my_mw_target3[,1])[my_mw_target3[,1]]
my_mw_target4 &lt;- cbind(my_mw_target3, Freq=mycounts[as.character(my_mw_target3[,1])]) 
</code></pre></div></div>

<ul>
  <li><strong>Step 8</strong>: Perform a vectorized devision of columns 3 and 4
  (average AA weight per protein)</li>
</ul>

<!-- -->

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>data.frame(my_mw_target4, avg_AA_WT=(my_mw_target4[,3] / my_mw_target4[,4]))[1:2,5:11] 

##   Loc  cTP   mTP    SP other Freq.Var1 Freq.Freq
## 1   _ 0.10 0.090 0.075 0.925 AT1G01010         1
## 2   * 0.01 0.636 0.158 0.448 AT1G01020         2
</code></pre></div></div>

<ul>
  <li><strong>Step 9</strong>: Calculate for each row the mean and standard deviation
  across several columns</li>
</ul>

<!-- -->

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mymean &lt;- apply(my_mw_target4[,6:9], 1, mean)
mystdev &lt;- apply(my_mw_target4[,6:9], 1, sd, na.rm=TRUE)
data.frame(my_mw_target4, mean=mymean, stdev=mystdev)[1:2,5:12] 

##   Loc  cTP   mTP    SP other Freq.Var1 Freq.Freq   mean
## 1   _ 0.10 0.090 0.075 0.925 AT1G01010         1 0.2975
## 2   * 0.01 0.636 0.158 0.448 AT1G01020         2 0.3130
</code></pre></div></div>

<h3 id="plotting-example">Plotting Example</h3>

<ul>
  <li><strong>Step 10</strong>: Generate scatter plot columns: ‘MW’ and ‘Residues’</li>
</ul>

<!-- -->

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>plot(my_mw_target4[1:500,3:4], col="red")
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/plot_example-1.svg" alt="" /></p>

<h3 id="export-results-and-run-entire-exercise-as-script">Export Results and Run Entire Exercise as Script</h3>

<ul>
  <li><strong>Step 11</strong>: Write the data frame <code class="language-plaintext highlighter-rouge">my_mw_target4</code> into a
  tab-delimited text file and inspect it in Excel.</li>
</ul>

<!-- -->

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>write.table(my_mw_target4, file="my_file.xls", quote=F, sep="\t", col.names = NA) 
</code></pre></div></div>

<ul>
  <li><strong>Homework 3H</strong>: Write all commands from this exercise into an R
  script named <code class="language-plaintext highlighter-rouge">exerciseRbasics.R</code>, or download it from
  <a href="http://faculty.ucr.edu/~tgirke/Documents/R_BioCond/My_R_Scripts/exerciseRbasics.R">here</a>.
  Then execute the script with the <code class="language-plaintext highlighter-rouge">source</code> function like this:
  <code class="language-plaintext highlighter-rouge">source("exerciseRbasics.R")</code>. This will run all commands of this
  exercise and generate the corresponding output files in the current
  working directory.</li>
</ul>

<!-- -->

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>source("exerciseRbasics.R")
</code></pre></div></div>

<h2 id="r-markdown">R Markdown</h2>

<h3 id="overview-1">Overview</h3>

<p>R Markdown combines markdown (an easy to write plain text format) with
embedded R code chunks. When compiling R Markdown documents, the code
components can be evaluated so that both the code and its output can be
included in the final document. This makes analysis reports highly
reproducible by allowing to automatically regenerate them when the
underlying R code or data changes. R Markdown documents (<code class="language-plaintext highlighter-rouge">.Rmd</code> files)
can be rendered to various formats including HTML and PDF. The R code in
an <code class="language-plaintext highlighter-rouge">.Rmd</code> document is processed by <code class="language-plaintext highlighter-rouge">knitr</code>, while the resulting <code class="language-plaintext highlighter-rouge">.md</code>
file is rendered by <code class="language-plaintext highlighter-rouge">pandoc</code> to the final output formats (<em>e.g.</em> HTML or
PDF). Historically, R Markdown is an extension of the older
<code class="language-plaintext highlighter-rouge">Sweave/Latex</code> environment. Rendering of mathematical expressions and
reference management is also supported by R Markdown using embedded
Latex syntax and Bibtex, respectively.</p>

<h3 id="quick-start">Quick Start</h3>

<h4 id="install-r-markdown">Install R Markdown</h4>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>install.packages("rmarkdown")
</code></pre></div></div>

<h4 id="initialize-a-new-r-markdown-rmd-script">Initialize a new R Markdown (<code class="language-plaintext highlighter-rouge">Rmd</code>) script</h4>

<p>To minimize typing, it can be helful to start with an R Markdown
template and then modify it as needed. Note the file name of an R
Markdown scirpt needs to have the extension <code class="language-plaintext highlighter-rouge">.Rmd</code>. Template files for
the following examples are available here:</p>

<ul>
  <li>R Markdown sample script:
  <a href="https://raw.githubusercontent.com/tgirke/GEN242/gh-pages/_vignettes/07_Rbasics/sample.Rmd"><code class="language-plaintext highlighter-rouge">sample.Rmd</code></a></li>
  <li>Bibtex file for handling citations and reference section:
  <a href="https://raw.githubusercontent.com/tgirke/GEN242/gh-pages/_vignettes/07_Rbasics/bibtex.bib"><code class="language-plaintext highlighter-rouge">bibtex.bib</code></a></li>
</ul>

<p>Users want to download these files, open the <code class="language-plaintext highlighter-rouge">sample.Rmd</code> file with
their preferred R IDE (<em>e.g.</em> RStudio, vim or emacs), initilize an R
session and then direct their R session to the location of these two
files.</p>

<h4 id="metadata-section">Metadata section</h4>

<p>The metadata section (YAML header) in an R Markdown script defines how
it will be processed and rendered. The metadata section also includes
both title, author, and date information as well as options for
customizing the output format. For instance, PDF and HTML output can be
defined with <code class="language-plaintext highlighter-rouge">pdf_document</code> and <code class="language-plaintext highlighter-rouge">html_document</code>, respectively. The
<code class="language-plaintext highlighter-rouge">BiocStyle::</code> prefix will use the formatting style of the
<a href="http://bioconductor.org/packages/release/bioc/html/BiocStyle.html"><code class="language-plaintext highlighter-rouge">BiocStyle</code></a>
package from Bioconductor.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code> ---
title: "My First R Markdown Document"
author: "Author: First Last"
date: "Last update: 15 April, 2020"
output:
  BiocStyle::html_document:
    toc: true
    toc_depth: 3
    fig_caption: yes

fontsize: 14pt
bibliography: bibtex.bib
 ---
</code></pre></div></div>

<h4 id="render-rmd-script">Render <code class="language-plaintext highlighter-rouge">Rmd</code> script</h4>

<p>An R Markdown script can be evaluated and rendered with the following
<code class="language-plaintext highlighter-rouge">render</code> command or by pressing the <code class="language-plaintext highlighter-rouge">knit</code> button in RStudio. The
<code class="language-plaintext highlighter-rouge">output_format</code> argument defines the format of the output (<em>e.g.</em>
<code class="language-plaintext highlighter-rouge">html_document</code>). The setting <code class="language-plaintext highlighter-rouge">output_format="all"</code> will generate all
supported output formats. Alternatively, one can specify several output
formats in the metadata section as shown in the above example.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>rmarkdown::render("sample.Rmd", clean=TRUE, output_format="html_document")
</code></pre></div></div>

<p>The following shows two options how to run the rendering from the
command-line.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ Rscript -e "rmarkdown::render('sample.Rmd', clean=TRUE)"
</code></pre></div></div>

<p>Alternatively, one can use a Makefile to evaluate and render an R
Markdown script. A sample Makefile for rendering the above <code class="language-plaintext highlighter-rouge">sample.Rmd</code>
can be downloaded
<a href="https://raw.githubusercontent.com/tgirke/GEN242/gh-pages/_vignettes/07_Rbasics/Makefile"><code class="language-plaintext highlighter-rouge">here</code></a>.
To apply it to a custom <code class="language-plaintext highlighter-rouge">Rmd</code> file, one needs open the Makefile in a
text editor and change the value assigned to <code class="language-plaintext highlighter-rouge">MAIN</code> (line 13) to the
base name of the corresponding <code class="language-plaintext highlighter-rouge">.Rmd</code> file (<em>e.g.</em> assign
<code class="language-plaintext highlighter-rouge">systemPipeRNAseq</code> if the file name is <code class="language-plaintext highlighter-rouge">systemPipeRNAseq.Rmd</code>). To
execute the <code class="language-plaintext highlighter-rouge">Makefile</code>, run the following command from the command-line.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>$ make -B
</code></pre></div></div>

<h4 id="r-code-chunks">R code chunks</h4>

<p>R Code Chunks can be embedded in an R Markdown script by using three
backticks at the beginning of a new line along with arguments enclosed
in curly braces controlling the behavior of the code. The following
lines contain the plain R code. A code chunk is terminated by a new line
starting with three backticks. The following shows an example of such a
code chunk. Note the backslashes are not part of it. They have been
added to print the code chunk syntax in this document.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>    ```\{r code_chunk_name, eval=FALSE\}
    x &lt;- 1:10
    ```
</code></pre></div></div>

<p>The following lists the most important arguments to control the behavior
of R code chunks:</p>

<ul>
  <li><code class="language-plaintext highlighter-rouge">r</code>: specifies language for code chunk, here R</li>
  <li><code class="language-plaintext highlighter-rouge">chode_chunk_name</code>: name of code chunk; this name needs to be unique</li>
  <li><code class="language-plaintext highlighter-rouge">eval</code>: if assigned <code class="language-plaintext highlighter-rouge">TRUE</code> the code will be evaluated</li>
  <li><code class="language-plaintext highlighter-rouge">warning</code>: if assigned <code class="language-plaintext highlighter-rouge">FALSE</code> warnings will not be shown</li>
  <li><code class="language-plaintext highlighter-rouge">message</code>: if assigned <code class="language-plaintext highlighter-rouge">FALSE</code> messages will not be shown</li>
  <li><code class="language-plaintext highlighter-rouge">cache</code>: if assigned <code class="language-plaintext highlighter-rouge">TRUE</code> results will be cached to reuse in
  future rendering instances</li>
  <li><code class="language-plaintext highlighter-rouge">fig.height</code>: allows to specify height of figures in inches</li>
  <li><code class="language-plaintext highlighter-rouge">fig.width</code>: allows to specify width of figures in inches</li>
</ul>

<p>For more details on code chunk options see
<a href="https://www.rstudio.com/wp-content/uploads/2015/03/rmarkdown-reference.pdf">here</a>.</p>

<h4 id="learning-markdown">Learning Markdown</h4>

<p>The basic syntax of Markdown and derivatives like kramdown is extremely
easy to learn. Rather than providing another introduction on this topic,
here are some useful sites for learning Markdown:</p>

<ul>
  <li><a href="https://guides.github.com/features/mastering-markdown/">Markdown Intro on
  GitHub</a></li>
  <li><a href="https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet">Markdown Cheet
  Sheet</a></li>
  <li><a href="http://rmarkdown.rstudio.com/authoring_basics.html">Markdown Basics from
  RStudio</a></li>
  <li><a href="http://www.rstudio.com/wp-content/uploads/2015/02/rmarkdown-cheatsheet.pdf">R Markdown Cheat
  Sheet</a></li>
  <li><a href="http://kramdown.gettalong.org/syntax.html">kramdown Syntax</a></li>
</ul>

<h4 id="tables">Tables</h4>

<p>There are several ways to render tables. First, they can be printed
within the R code chunks. Second, much nicer formatted tables can be
generated with the functions <code class="language-plaintext highlighter-rouge">kable</code>, <code class="language-plaintext highlighter-rouge">pander</code> or <code class="language-plaintext highlighter-rouge">xtable</code>. The
following example uses <code class="language-plaintext highlighter-rouge">kable</code> from the <code class="language-plaintext highlighter-rouge">knitr</code> package.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library(knitr)
kable(iris[1:12,])
</code></pre></div></div>

<table>
<thead>
<tr class="header">
<th style="text-align: right;">Sepal.Length</th>
<th style="text-align: right;">Sepal.Width</th>
<th style="text-align: right;">Petal.Length</th>
<th style="text-align: right;">Petal.Width</th>
<th style="text-align: left;">Species</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: right;">5.1</td>
<td style="text-align: right;">3.5</td>
<td style="text-align: right;">1.4</td>
<td style="text-align: right;">0.2</td>
<td style="text-align: left;">setosa</td>
</tr>
<tr class="even">
<td style="text-align: right;">4.9</td>
<td style="text-align: right;">3.0</td>
<td style="text-align: right;">1.4</td>
<td style="text-align: right;">0.2</td>
<td style="text-align: left;">setosa</td>
</tr>
<tr class="odd">
<td style="text-align: right;">4.7</td>
<td style="text-align: right;">3.2</td>
<td style="text-align: right;">1.3</td>
<td style="text-align: right;">0.2</td>
<td style="text-align: left;">setosa</td>
</tr>
<tr class="even">
<td style="text-align: right;">4.6</td>
<td style="text-align: right;">3.1</td>
<td style="text-align: right;">1.5</td>
<td style="text-align: right;">0.2</td>
<td style="text-align: left;">setosa</td>
</tr>
<tr class="odd">
<td style="text-align: right;">5.0</td>
<td style="text-align: right;">3.6</td>
<td style="text-align: right;">1.4</td>
<td style="text-align: right;">0.2</td>
<td style="text-align: left;">setosa</td>
</tr>
<tr class="even">
<td style="text-align: right;">5.4</td>
<td style="text-align: right;">3.9</td>
<td style="text-align: right;">1.7</td>
<td style="text-align: right;">0.4</td>
<td style="text-align: left;">setosa</td>
</tr>
<tr class="odd">
<td style="text-align: right;">4.6</td>
<td style="text-align: right;">3.4</td>
<td style="text-align: right;">1.4</td>
<td style="text-align: right;">0.3</td>
<td style="text-align: left;">setosa</td>
</tr>
<tr class="even">
<td style="text-align: right;">5.0</td>
<td style="text-align: right;">3.4</td>
<td style="text-align: right;">1.5</td>
<td style="text-align: right;">0.2</td>
<td style="text-align: left;">setosa</td>
</tr>
<tr class="odd">
<td style="text-align: right;">4.4</td>
<td style="text-align: right;">2.9</td>
<td style="text-align: right;">1.4</td>
<td style="text-align: right;">0.2</td>
<td style="text-align: left;">setosa</td>
</tr>
<tr class="even">
<td style="text-align: right;">4.9</td>
<td style="text-align: right;">3.1</td>
<td style="text-align: right;">1.5</td>
<td style="text-align: right;">0.1</td>
<td style="text-align: left;">setosa</td>
</tr>
<tr class="odd">
<td style="text-align: right;">5.4</td>
<td style="text-align: right;">3.7</td>
<td style="text-align: right;">1.5</td>
<td style="text-align: right;">0.2</td>
<td style="text-align: left;">setosa</td>
</tr>
<tr class="even">
<td style="text-align: right;">4.8</td>
<td style="text-align: right;">3.4</td>
<td style="text-align: right;">1.6</td>
<td style="text-align: right;">0.2</td>
<td style="text-align: left;">setosa</td>
</tr>
</tbody>
</table>

<p>A much more elegant and powerful solution is to create fully interactive
tables with the <a href="https://rstudio.github.io/DT/"><code class="language-plaintext highlighter-rouge">DT</code> package</a>. This
JavaScirpt based environment provides a wrapper to the DataTables
library using jQuery. The resulting tables can be sorted, queried and
resized by the user. For an example see
<a href="http://girke.bioinformatics.ucr.edu/GEN242/pages/mydoc/Rbasics.html#1627_tables">here</a>.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library(DT)
datatable(iris, filter = 'top', options = list(
      pageLength = 100, scrollX = TRUE, scrollY = "600px", autoWidth = TRUE
     ))
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/dt-1.png" alt="" /></p>

<h4 id="figures">Figures</h4>

<p>Plots generated by the R code chunks in an R Markdown document can be
automatically inserted in the output file. The size of the figure can be
controlled with the <code class="language-plaintext highlighter-rouge">fig.height</code> and <code class="language-plaintext highlighter-rouge">fig.width</code> arguments.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library(ggplot2)
dsmall &lt;- diamonds[sample(nrow(diamonds), 1000), ]
ggplot(dsmall, aes(color, price/carat)) + geom_jitter(alpha = I(1 / 2), aes(color=color))
</code></pre></div></div>

<p><img src="gen242_files/figure-markdown_strict/some_jitter_plot-1.svg" alt="" /></p>

<p>Sometimes it can be useful to explicitly write an image to a file and
then insert that image into the final document by referencing its file
name in the R Markdown source. For instance, this can be useful for time
consuming analyses. The following code will generate a file named
<code class="language-plaintext highlighter-rouge">myplot.png</code>. To insert the file in the final document, one can use
standard Markdown or HTML syntax.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>dev.off()

## null device 
##           1
</code></pre></div></div>

<h4 id="inline-r-code">Inline R code</h4>

<p>To evaluate R code inline, one can enclose an R expression with a single
back-tick followed by <code class="language-plaintext highlighter-rouge">r</code> and then the actual expression. For instance,
the back-ticked version of ‘r 1 + 1’ evaluates to 2 and ‘r pi’ evaluates
to 3.1415927.</p>

<h4 id="mathematical-equations">Mathematical equations</h4>

<p>To render mathematical equations, one can use standard Latex syntax.
When expressions are enclosed with single <code class="language-plaintext highlighter-rouge">$</code> signs then they will be
shown inline, while enclosing them with double <code class="language-plaintext highlighter-rouge">$$</code> signs will show them
in display mode. For instance, the following Latex syntax
<code class="language-plaintext highlighter-rouge">d(X,Y) = \sqrt[]{ \sum_{i=1}^{n}{(x_{i}-y_{i})^2} }</code> renders in display
mode as follows:</p>

<script type="math/tex; mode=display">d(X,Y) = \\sqrt\[\]{ \\sum\_{i=1}^{n}{(x\_{i}-y\_{i})^2} }</script>

<h4 id="citations-and-bibliographies">Citations and bibliographies</h4>

<p>Citations and bibliographies can be autogenerated in R Markdown in a
similar way as in Latex/Bibtex. Reference collections should be stored
in a separate file in Bibtex or other supported formats. To cite a
publication in an R Markdown script, one uses the syntax <code class="language-plaintext highlighter-rouge">[@&lt;id1&gt;]</code>
where <code class="language-plaintext highlighter-rouge">&lt;id1&gt;</code> needs to be replaced with a reference identifier present
in the Bibtex database listed in the metadata section of the R Markdown
script (<em>e.g.</em> <code class="language-plaintext highlighter-rouge">bibtex.bib</code>). For instance, to cite Lawrence et al.
(2013), one uses its reference identifier (<em>e.g.</em> <code class="language-plaintext highlighter-rouge">Lawrence2013-kt</code>) as
<code class="language-plaintext highlighter-rouge">&lt;id1&gt;</code> [@Lawrence2013-kt]. This will place the citation inline in the
text and add the corresponding reference to a reference list at the end
of the output document. For the latter a special section called
<code class="language-plaintext highlighter-rouge">References</code> needs to be specified at the end of the R Markdown script.
To fine control the formatting of citations and reference lists, users
want to consult this the corresponding <a href="http://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html">R Markdown
page</a>.
Also, for general reference management and outputting references in
Bibtex format <a href="https://paperpile.com/features">Paperpile</a> can be very
helpful.</p>

<h4 id="viewing-r-markdown-report-on-hpcc-cluster">Viewing R Markdown report on HPCC cluster</h4>

<p>R Markdown reports located on UCR’s HPCC Cluster can be viewed locally
in a web browser (without moving the source HTML) by creating a symbolic
link from a user’s <code class="language-plaintext highlighter-rouge">.html</code> directory. This way any updates to the report
can be viewed immediately without creating another copy of the HTML
file. For instance, if user <code class="language-plaintext highlighter-rouge">ttest</code> has generated an R Markdown report
under <code class="language-plaintext highlighter-rouge">~/bigdata/today/rmarkdown/sample.html</code>, then the proper symbolic
link to this file can be created as follows:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>cd ~/.html
ln -s ~/bigdata/today/rmarkdown/sample.html sample.html
</code></pre></div></div>

<p>After this one can view the report in a web browser using this URL
<a href="http://biocluster.ucr.edu/~ttest/rmarkdown/sample.html">http://biocluster.ucr.edu/~ttest/rmarkdown/sample.html</a>. If necessary
access to the URL can be restricted with a password following the
instructions
<a href="http://hpcc.ucr.edu/manuals_linux-cluster_sharing.html#sharing-files-on-the-web">here</a>.</p>

<p>A sample R Markdown report for an RNA-Seq project is given here:</p>

<ul>
  <li><a href="https://girke.bioinformatics.ucr.edu/GEN242/pages/mydoc/demo_files/systemPipeRNAseq.html" target="\_blank">RNASeq.html</a></li>
  <li><a href="https://raw.githubusercontent.com/tgirke/GEN242/gh-pages/pages/mydoc/demo_files/systemPipeRNAseq.Rmd" target="\_blank">RNASeq.Rmd</a></li>
</ul>

<h2 id="shiny-web-apps">Shiny Web Apps</h2>

<h3 id="what-is-shiny">What is Shiny?</h3>

<p><a href="https://shiny.rstudio.com/gallery/">Shiny</a> is an R-based environment
for building interactive web applications for data analysis and
exploration. Since most JavaScript code is autogenerated by the
environment, basic R knowledge is sufficient for developing Shiny apps.
They can be deployed on local computers or web servers including custom
and cloud-based servers ( e.g. AWS, GCP,
<a href="http://www.shinyapps.io/">shinyapp.io</a> service). The basic structure of
a Shiny app is an <code class="language-plaintext highlighter-rouge">app.R</code> script containing the following components:</p>

<ol>
  <li>
    <p>User interface</p>

    <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>ui &lt;- fluidPage()
</code></pre></div>    </div>
  </li>
  <li>
    <p>Server function</p>

    <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>server &lt;- function(input, output) {}
</code></pre></div>    </div>
  </li>
  <li>
    <p>Statement to run shiny app</p>

    <div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>shinyApp(ui = ui, server = server)
</code></pre></div>    </div>
  </li>
</ol>

<p>Alternatively, the <code class="language-plaintext highlighter-rouge">ui</code> and <code class="language-plaintext highlighter-rouge">server</code> functions can be organized in two
script files, a <code class="language-plaintext highlighter-rouge">ui.R</code> and a <code class="language-plaintext highlighter-rouge">server.R</code> script, respectively.</p>

<h3 id="develop-and-test-shiny-app-locally">Develop and test Shiny app locally</h3>

<p>Open R and set session to parent directory (here <code class="language-plaintext highlighter-rouge">myappdir</code>) containing
shiny script <code class="language-plaintext highlighter-rouge">app.R</code>, and the run it with the <code class="language-plaintext highlighter-rouge">runApp()</code> function. A
sample <code class="language-plaintext highlighter-rouge">app.R</code> script for testing can be downloaded from
<a href="https://raw.githubusercontent.com/tgirke/GEN242/gh-pages/_vignettes/07_Rbasics/shinyapp/app.R">here</a>.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>library(shiny)
runApp("myappdir") # To show code in app, add argument: display.mode="showcase" 
</code></pre></div></div>

<p>This will open the app in a web browser.</p>

<h3 id="deploy-on-web-server">Deploy on web server</h3>

<p>This can be done on local or cloud systems. An easy solution is to get
an account on <a href="http://www.shinyapps.io/">shinyapps.io</a> and then deploy
Shiny apps there. For details, see
<a href="https://shiny.rstudio.com/deploy/">here</a>.</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>setwd("myappdir")
library(rsconnect)
deployApp()
</code></pre></div></div>

<h3 id="example-shiny-app">Example Shiny app</h3>

<p>The following Shiny app is hosted on <code class="language-plaintext highlighter-rouge">shinyapps.io</code> and embedded into
the markdown (or html) source of this page using the following iframe
syntax:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>&lt;iframe src="https://tgirke.shinyapps.io/diamonds/" style="border: none; width: 880px; height: 900px"&gt;&lt;/iframe&gt;
</code></pre></div></div>

<iframe src="https://tgirke.shinyapps.io/diamonds/" style="border: none; width: 880px; height: 900px">
</iframe>

<h3 id="learning-shiny">Learning Shiny</h3>

<p>The Shiny section on the Rstudio site contains excellent
<a href="https://shiny.rstudio.com/tutorial/">tutorials</a>. In addition, users may
want to explore the example apps included in the <code class="language-plaintext highlighter-rouge">shiny</code> package. This
can be done by loading the individual examples (see
<a href="https://shiny.rstudio.com/tutorial/written-tutorial/lesson1/">here</a>) or
saving the code to a user writable directory like this:</p>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>mydir &lt;- system.file("examples", package="shiny")
dir.create('my_shiny_test_dir')
file.copy(mydir, "my_shiny_test_dir", recursive=TRUE)
setwd("my_shiny_test_dir/examples")
runApp("01_hello") # Runs first example app in directory 
dir() # Lists available Shiny examples (directories). 
</code></pre></div></div>

<h2 id="session-info">Session Info</h2>

<div class="language-plaintext highlighter-rouge"><div class="highlight"><pre class="highlight"><code>sessionInfo()
</code></pre></div></div>

<h2 id="references">References</h2>

<h1 id="intro-r">Intro R</h1>
<p>http://www.r-tutor.com/r-introduction/data-frame</p>

<p>https://github.com/tgirke/GEN242/blob/gh-pages/_vignettes/07_Rbasics/Rbasics.Rmd</p>

<h2 id="books-and-documentation">Books and Documentation</h2>

<ul>
  <li>simpleR - Using R for Introductory Statistics (John Verzani, 2004) -
  <a href="http://cran.r-project.org/doc/contrib/Verzani-SimpleR.pdf">URL</a></li>
  <li>Bioinformatics and Computational Biology Solutions Using R and
  Bioconductor (Gentleman et al., 2005) -
  <a href="http://www.bioconductor.org/help/publications/books/bioinformatics-and-computational-biology-solutions/">URL</a></li>
  <li>More on this see “Finding Help” section in UCR Manual -
  <a href="http://manuals.bioinformatics.ucr.edu/home/R_BioCondManual#TOC-Finding-Help">URL</a></li>
</ul>

						</div><!-- /.content -->
					</div><!-- /.col -->
					<div class="col-md-4 col-md-offset-1">
						<div class="sections-list-wrapper">
							<div class="sections-list js-sections js-affix js-scrollspy hidden-xs hidden-sm"></div><!-- /.sections-list -->
						</div>
					</div><!-- /.col -->
				</div><!-- /.row -->
			</div><!-- /.container -->
		</div><!-- /.section -->
		
		<div class="js-footer-area">
			
				<nav class="page-nav">
					<div class="container">
						<div class="row">
							<div class="col-xs-12">
								
									<a href="/u3/" class="page-nav__item page-nav__item--prev">
										<i class="icon icon--arrow-left"></i>
										Project Management - Git
									</a><!-- /.page-nav__item -->
								
								
									<a href="/u4/help/" class="page-nav__item page-nav__item--next">
										Help
										<i class="icon icon--arrow-right"></i>
									</a><!-- /.page-nav__item -->
								
							</div><!-- /.col -->
						</div><!-- /.row -->
					</div><!-- /.container -->
				</nav><!-- /.page-nav -->
			
			
			
	<footer class="site-footer">
		<div class="container">
			<div class="row">
				<div class="col-sm-6">
					
						<a href="/" class="site-footer__logo">R for Plant Physiologists</a>
					
					
						<hr>
						<p class="site-footer__copyright">Contact: <a href="mailto:danicassol@gmail.com">danicassol@gmail.com</a> </br> ©2020 Daniela Cassol. All rights reserved.</p>
					
				</div><!-- /.col -->
				
					<div class="col-sm-6 align-right">
						<ul class="social-list">
							
								<li>
									<a href="https://www.github.com/dcassol" target="_blank" class="social-list__item social-list__item--github">
										<i class="icon icon--github"></i>
									</a>
								</li>
							
						</ul><!-- /.social-list -->
					</div><!-- /.col -->
				
			</div><!-- /.row -->
		</div><!-- /.container -->
	</footer><!-- /.site-footer -->


<script src="/doks-theme/assets/js/vendor/jquery.min.js"></script>
<script type="text/javascript" src="/doks-theme/assets/js/vendor/bootstrap/affix.min.js"></script>
<script type="text/javascript" src="/doks-theme/assets/js/vendor/bootstrap/scrollspy.min.js"></script>
<script type="text/javascript" src="/doks-theme/assets/js/vendor/matchHeight.min.js"></script>
<script type="text/javascript" src="/doks-theme/assets/js/scripts.min.js"></script>




		</div><!-- /.js-footer-area -->
	</body>
</html>
